---
title: 'Feature selection of cell types in scRNAseq Tabula sapiens (Blood tissue)
  V2 '
author: "Antonin Colajanni"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_depth: 2
    toc_float:
      collapsed: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

# Tabula Sapiens Analyses : RF feature reduction

```{r  fix_knit, fig.width=12, fig.height=9, eval=TRUE, include=FALSE,warning=FALSE , message=FALSE , message=FALSE}
library(rmarkdown)

# setwd("/home/acolajanni/Documents/work/results/Tabula_sapiens/markdown/")
# clean_tmpfiles_mod <- function() {
#   message("Calling clean_tmpfiles_mod()")
# }
# 
# assignInNamespace("clean_tmpfiles", clean_tmpfiles_mod, ns = "rmarkdown")

library(purrr)
rate <- rate_backoff(pause_base = 0.1, pause_min = 0.005, max_times = 10)
insistent_render <- insistently(rmarkdown::render, rate, quiet = FALSE)


```

```{r import, echo=T, results='hide', message=F, warning=F, eval=T}

# I M P O R T
library(stringr)
library (plyr) 
library(ggplot2)
library(viridis)
library(UpSetR)
library(ggrepel)
library(factoextra)
library(FactoMineR)
library(viridis)
library(RColorBrewer)
library(dplyr)
library(pheatmap)
library(DT) 
library(cluster)
library(grid)
# V A R I A B L E S
#main.dir = "/run/user/1001/gvfs/sftp:host=curta.mcia.fr,user=acolajanni/gpfs/home/acolajanni"
path = "/run/user/1001/gvfs/sftp:host=core.cluster.france-bioinformatique.fr,user=acolajanni/shared/projects/microbiome_translocation"
local_path = "/home/acolajanni/Documents/work/"
#setwd(main.dir)

#path = "/home/acolajanni/Documents/work/"
data_dir = file.path(path, "data/Tabula_sapiens_immune_all")
result_dir = file.path(path, "results/Tabula_sapiens/RF_filter_FC_v2/")
#setwd(path)

git_dir = file.path(local_path,"GITHUB/Signature_matrix_TabulaSapiens/")
git_geneset = file.path(git_dir,"genesets/")
################################################################################
# F U N C T I O N S

create_dir = function(directory){
  if ( ! file.exists(directory)){
    dir.create(directory)  }
  return(directory) }

replace_label = function(label, name_mapping){
  
  label_names = names(name_mapping)
  label = ifelse(str_detect(label, label_names[1]),name_mapping[1],label)
  label = ifelse(str_detect(label,label_names[2]),name_mapping[2],label)
  label = ifelse(label %in% c(name_mapping[1], name_mapping[2]), yes = label, no = name_mapping[3])
  #label=as.numeric(label)
  return(label) }


confidence_interval = function(perm_df, n_perm = 20) {
  t.score=qt(p=0.025, df=n_perm-1 ,lower.tail = FALSE)
  perm_df$se = perm_df$std_importance/sqrt(n_perm-1)
  perm_df$margin_error = t.score*perm_df$se
  perm_df$binf = perm_df$mean_importance - perm_df$margin_error
  perm_df$bsup = perm_df$mean_importance + perm_df$margin_error
  
  return(perm_df) }


importance_barplot = function(permutation_df,permutation_number=50, filter="mean", title="Mean feature importance barplot",
                              conf_interval = TRUE){
  if (conf_interval){
    permutation_df = confidence_interval(permutation_df, permutation_number) 
  }
  else {
    permutation_df$binf = permutation_df$mean_importance
    permutation_df$bsup = permutation_df$mean_importance
  }
  non_null_feature = length(permutation_df[permutation_df$mean_importance != 0 , ]$mean_importance)
  pos_feature = length(permutation_df[permutation_df$mean_importance > 0 , ]$mean_importance)

  
  if (filter == "mean") { 
    permutation_df = permutation_df[permutation_df$mean_importance > 0 , ]
    subtitle = paste(non_null_feature, "Genes with non null feature importance among which", pos_feature, "have a positive importance") }
  else {  
    sig_pos_feature = length(permutation_df[permutation_df$binf > 0 , ]$mean_importance)
    subtitle = paste(non_null_feature, "Genes with non null feature importance among which", pos_feature, "have a positive importance and \n ",
                     sig_pos_feature, "are significantly different than zero.")
    
    if (filter == "bsup") { permutation_df = permutation_df[permutation_df$bsup > 0 , ] }
    else if (filter == "binf") { permutation_df = permutation_df[permutation_df$binf > 0 , ] }
    else if (filter == "zero") { permutation_df = permutation_df[permutation_df$mean_importance != 0 , ] }
  }
  
  
  permutation_df$genes = row.names(permutation_df)
  n_gene = length(row.names(permutation_df))
  ordered_permutation_df = permutation_df[order(permutation_df$mean_importance, decreasing = TRUE),]
  
  plot = ggplot( permutation_df, aes(x = reorder(genes,mean_importance), y = mean_importance, fill=mean_importance ) )  +
    geom_col(position = position_dodge(0), width = 0.75) + 
    geom_errorbar(aes(ymin=binf, ymax=bsup ),
                size=.3,width=.2, position=position_dodge(.9)) +
    geom_hline(aes(yintercept = 0), alpha=0.20) +
    coord_flip() +
    labs(fill = "Cluster") +
    ylab('Mean Permutation Feature Importance') + xlab('Gene name')+
    ggtitle(title, subtitle = subtitle ) +
  
    scale_fill_viridis(discrete = FALSE, alpha=0.65, direction = -1, option='C', name="Mean Feature Importance") +
    theme_linedraw()+  
    theme( plot.title = element_text(size=15),
           axis.text.y = element_text(size = 8))
  return(list("plot"=plot, 
              "ordered_permutation_dataframe" = ordered_permutation_df[,-c(1:permutation_number)],
              "permutation_dataframe"= permutation_df)) }


add_annotation_line_barplot_importance = function(plot, permutation_df, last_gene, annotation = "most important features", adjust_annotation = .25, adjust_annotationX=0.85){
  gene_value = permutation_df[permutation_df$genes == last_gene,]$mean_importance
  gene_rank = length(permutation_df[permutation_df$mean_importance >= gene_value,]$genes)
  
  plot = plot +
    geom_vline( xintercept = which(permutation_df$genes == last_gene)-0.5, linetype='dashed' )+
    annotate("text", x=which(permutation_df$genes == last_gene)+adjust_annotation , y=max(permutation_df$mean_importance)*adjust_annotationX,
             label= paste(gene_rank,annotation) )
    return(plot) }

RA_plot = function(FC_df, x, y, title = "RA plot", highlight=c('TOP20'), annotation = "Most Important Features", vjust = 2, force=3, y_scale = 0, add_label=TRUE){
  
  FC_df$selection = ifelse(FC_df[[x]] > quantile(FC_df[[x]],0.75) & abs(FC_df[[y]]) > 1 , 
                            yes = "Selected",no = "Not selected")
  selected_number= length(FC_df$selection[FC_df$selection == "Selected"])
    
  if (highlight[1] == "TOP20") {
    FC_df$selection = ifelse(abs(FC_df[[y]]) > sort(abs(FC_df[[y]]),TRUE )[21]  , 
                            yes = "20 Highest FC",no = FC_df$selection)
    FC_df$label = ifelse(FC_df$selection == "20 Highest FC", 
                          yes=row.names(FC_df), no="") }
  else{
    FC_df$selection = ifelse(row.names(FC_df) %in% highlight , 
                            yes = annotation, no = FC_df$selection)
    FC_df$label = ifelse(FC_df$selection == annotation, 
                          yes=row.names(FC_df), no="") }
  
  
  maximum_y = max(abs(FC_df[[y]])) + y_scale
  title =paste( c(title, "\n Gene selected :",selected_number),collapse = " " )
  
  plot = ggplot(FC_df, aes(x=FC_df[[x]], y=FC_df[[y]]))+
    geom_point(aes(color=selection, alpha=selection, size = selection))+
    geom_smooth(color="darkred",size=0.5)+
    annotate("text", x=2.4, y=-1-0.5, label="log2 FC = -1", colour = "darkblue") +
    geom_hline(yintercept = -1, color='darkblue',linetype='dashed') +
    annotate("text", x=2.4, y= 1+0.5, label="log2 FC = 1", colour = "darkblue") +
    coord_cartesian(ylim=c(-maximum_y,maximum_y))+
    geom_hline(yintercept = 1, color='darkblue',linetype='dashed') +
    geom_vline(xintercept = quantile(FC_df[[x]],0.75),linetype='dashed') +
    annotate("text", y=-5.5, x= quantile(FC_df[[x]],0.75)/4, 
           label="3rd quartile \n of mean expression", colour = "black") +
    
    scale_color_manual(values = c("orangered","#777777","#00539CFF"))+
    scale_alpha_manual(values=c(1,0.25, 0.75), guide='none')+
    scale_size_manual(values=c(2,0.5,1), guide='none')+
    theme_linedraw()+
    theme(legend.position = c(0.85, 0.15) ,legend.direction = 'vertical') +
    labs(color="Features", )+
    xlab("Average log expression") +
    ylab("log ratio of expression") +
    ggtitle(title)
  
  if (add_label){
    plot = plot +
      geom_text_repel(data=FC_df, label = FC_df$label,
                    vjust = vjust, force=force, max.overlaps = 100) }
  
  return(list("plot"=plot, "df"=FC_df)) }


visualize_annotation = function(Annotation_table, celltype){
  Freq_annot = as.data.frame(table(Annotation_table$goslim_goa_description))

  plot = ggplot(Freq_annot, aes(x = reorder(Var1,Freq), y = Freq, fill=Freq)) + 
          geom_col() + 
          coord_flip() +
          ggtitle(paste("Frequence of annotation terms for", celltype ,"geneset"))+
          xlab('GOSLIM GOA annotation terms') + ylab('Frequence')+ labs(fill="Frequence") +
          theme_linedraw() +
          theme(axis.text = element_text(size=12, face = "bold"))
  
  df = rmarkdown::paged_table(unique(Annotation_table[,c("hgnc_symbol","entrezgene_description")])  ) %>% datatable 
  
  return(list("df" = df, "plot"=plot)) }

geneset_to_binary_df = function(geneset_list){
  geneset_list_df = lapply(geneset_list, function(x) as.data.frame(x))
  tmp = ldply(geneset_list_df)
  tmp = reshape2::dcast(tmp, x ~ .id)
  row.names(tmp) = tmp$x
  tmp$x = NULL
  tmp = ifelse(is.na(tmp), 0,1)   
  return( as.data.frame(tmp) ) }

heatmap_from_geneset_list = function(geneset_list, tabula_column, title = " ", 
                                     min_representation = 2, cutree_rows = 1,
                                     clust=TRUE){
  
  n_cut = length(names(geneset_list))
  all_tabula = geneset_list[[tabula_column]]
  all_signature = unique(unlist(geneset_list[ names(geneset_list) != tabula_column ]))
  
  unique_tabula = all_tabula[! all_tabula %in% all_signature]
  Full_geneset = unique(unlist(geneset_list))
  geneset_list[["all"]] = Full_geneset
  df_heatmap = fromList(geneset_list)
  row.names(df_heatmap) = geneset_list$all 
  
  main = paste0(title , "\n", 
        "Presented genes appears in at least ", min_representation, 
        " signature matrix, or in Tabula Sapiens \n",
        length(unique_tabula), 
        " Features uniquely predicted in Tabula sapiens geneset " , 
        "\n 1 : Detected by signature matrix",
        "\n 0 : Not detected by signature matrix")
  
  if (! clust) {return(list("title"=main, "df" = df_heatmap))}
  
  pheatmap::pheatmap(df_heatmap[ 
    rowSums(df_heatmap) > min_representation | df_heatmap[[tabula_column]] == 1 , 
    colnames(df_heatmap) != "all"],
      clustering_distance_rows = "correlation", clustering_distance_cols = "correlation",
      cutree_cols = n_cut, cutree_rows = cutree_rows,
      color = c("#555555","red"), 
      na_col="lightgray",breaks = c(0, .5 , 1),
      fontsize_row = 14, fontsize_col = 15, angle_col = 0, legend_breaks = c(0,1),
      main = main )
}

FC_prop_expr = function(df, n_cell, max_cell){
  
  columns = colnames(df)[str_detect(colnames(df), "prop_expr")]
  A = columns[!str_detect(columns, "vs")]
  B = columns[ str_detect(columns, paste0("other")) ]

  df[[paste0("FC_",A)]] = compute_FC( df[[A]] , df[[B]] , 0)
  return(df) }

kneedle_threshold = function(df, y, sens = 1){
  df$value = abs(as.numeric(y))
  df = df[order(df$value, decreasing = TRUE),]
  df$rank = 1:length(y)
  elbow = kneedle::kneedle(x = df$rank, y = df$value, decreasing = TRUE, concave = FALSE, sensitivity = sens)
  
  plot = ggplot(df, aes(x=rank,y=value))+
    geom_line()+
    geom_hline(yintercept = elbow[2], color="blue",alpha=0.5) + geom_vline(xintercept = elbow[1], color="blue",alpha=0.5)+
    ylab("y value") + xlab("Gene Rank")
  
  return(list("rank" = elbow[1], "threshold" = elbow[2], "plot"=plot )) }

Rprop_plot = function(FC_df, FC_x, y, title = "plot", highlight=c('TOP20'), annotation = "Most Important Features", 
                      vjust = 2, force=3, x_scale = 0, add_label=TRUE, 
                      FC_threshold = 1, prop_threshold=0.25, FC_thresh2 = FC_threshold,
                      prop_threshold_to_show = prop_threshold) {
  
  FC_df$selection = ifelse(abs(FC_df[[FC_x]]) > FC_threshold & abs(FC_df[[y]]) > prop_threshold , 
                            yes = "Selected",no = "Not selected")
  Display_2nd_annotation = FALSE
  
  if (FC_threshold < FC_thresh2){
    FC_df$selection = ifelse(abs(FC_df[[FC_x]]) >= FC_thresh2, yes = "Selected", no = FC_df$selection ) 
    Display_2nd_annotation = TRUE }
  
  selected_number= length(FC_df$selection[FC_df$selection == "Selected"])
    
  if (highlight[1] == "TOP20") {
    
    tmp_df = FC_df[FC_df[[y]] > prop_threshold &  abs(FC_df[[FC_x]]) > FC_threshold , ]
    tmp_df = tmp_df[with(tmp_df, order(abs(tmp_df[[FC_x]]) , decreasing = TRUE)),]
     if ( length(row.names(tmp_df)) > 20) { selected = row.names(tmp_df)[1:20] } 
     else{  selected = row.names(tmp_df) }

    FC_df$selection = ifelse( row.names(FC_df) %in% selected, 
                            yes = "20 Highest FC expressed in most cells",no = FC_df$selection)
    FC_df$label = ifelse(FC_df$selection == "20 Highest FC expressed in most cells", 
                          yes=row.names(FC_df), no="") }
  
  if(highlight[1] == "selection"){
    FC_df$label = ifelse(FC_df$selection == "Selected", 
                          yes=row.names(FC_df), no="") } 
  else{
    FC_df$selection = ifelse(row.names(FC_df) %in% highlight , 
                            yes = annotation, no = FC_df$selection)
    FC_df$label = ifelse(FC_df$selection == annotation, 
                          yes=row.names(FC_df), no="") }
  
  
  maximum_x = max(abs(FC_df[[FC_x]])) + x_scale
  title =paste( c(title, "\n Gene selected :",selected_number),collapse = " " )
  
  plot = ggplot(FC_df, aes(x=FC_df[[FC_x]], y=FC_df[[y]]))+
    geom_point(aes(color=selection, alpha=selection, size = selection))+

    # Annotation of first filter
    annotate("text", y=-0.02, x=-FC_threshold-0.75, 
             label=paste0("log2 FC = -",FC_threshold), 
             colour = "darkblue") +
    geom_vline(xintercept = -FC_threshold, color='darkblue',linetype='dashed') +
    
    annotate("text", y=-0.02, x= FC_threshold+0.75, 
             label=paste0("log2 FC = ",FC_threshold), 
             colour = "darkblue") +
    geom_vline(xintercept = FC_threshold, color='darkblue',linetype='dashed') +
    # Proportion line
    geom_hline(yintercept = prop_threshold, color='darkblue',linetype='dashed') +

    
    coord_cartesian(xlim=c(-maximum_x,maximum_x), ylim=c(-0.025,1)) +
    
    scale_color_manual(values = c("orangered","#777777","#00539CFF"))+
    scale_alpha_manual(values=c(1,0.25, 0.75), guide='none')+
    scale_size_manual(values=c(2,0.5,1), guide='none')+
      
    theme_linedraw()+
    theme(legend.position = c(0.15, 0.85) ,legend.direction = 'vertical') +
    labs(color="Features")+
    ylab("Proportion of cells that express a gene") +
    xlab("log ratio of expression") +
    ggtitle(title)
    
  if (add_label){
    plot = plot +
      geom_text_repel(data=FC_df, label = FC_df$label,
                    vjust = vjust, force=force, max.overlaps = 100) }
      # Annotation of 2nd filter
  if (Display_2nd_annotation){
    plot = plot +
    annotate("text", y=-0.04, x=-FC_thresh2-1, 
             label=paste0("log2 FC = -",round(FC_thresh2, digits = 3)), 
             colour = '#555555') +
    geom_vline(xintercept = -FC_thresh2, color='#555555',linetype='dashed') +
    
    annotate("text", y=-0.04, x= FC_thresh2+1, 
             label=paste0("log2 FC = ",label=round(FC_thresh2, digits = 3)  ), 
             colour = '#555555') +
    geom_vline(xintercept = FC_thresh2, color='#555555',linetype='dashed')  }
  
  if (prop_threshold < 0.1) {
      plot = plot + annotate("text", y=0.05, x= maximum_x*0.75, 
             label=paste0("Proportion of \n cells expressing = ", prop_threshold_to_show), 
             colour = "darkblue") }
  else {
      plot = plot + annotate("text", y=prop_threshold*1.15, x= -maximum_x*0.75, 
             label=paste0("Proportion of cells expressing = ", round(prop_threshold, 3)), 
             colour = "darkblue") }
  
  return(list("plot"=plot, "df"=FC_df)) }


get_permutation_list = function(directory,pattern2,pattern1="Permutation_result_"){
  # Import files
  #directory = file.path(directory)
  tmp = list.files(directory, full.names = TRUE, 
                   recursive = TRUE, pattern=pattern1)

  if (pattern1 == "Permutation_result_") {
    Permutation_list = lapply(tmp, 
                            function(x) x = read.csv2(x, header = TRUE, 
                                            sep = ",", row.names = 1))}
  else {
    Permutation_list = lapply(tmp, function(x) {
      x = read.table(x, header = TRUE , sep=",", row.names=2) } )} 
  
  names(Permutation_list) = tmp
  # Rename each dataframe
  names(Permutation_list) = str_remove_all(
    string = names(Permutation_list) , 
    pattern = paste0(directory , pattern2))
  names(Permutation_list) = str_remove_all(string = names(Permutation_list) , 
                                           pattern =".csv")
  # Change data to numeric
  Permutation_list = lapply(Permutation_list, function(df) {
    df = dplyr::mutate_all(df, function(x) as.numeric(as.character(x) )) } )
  
  # return(list('l'=0,
  #             "dir"=paste0(directory , pattern2),
  #             "names"=names(Permutation_list) ) ) }
  return(Permutation_list)  }


get_genes_from_binary=function(binary_df, expressed_genes=NULL, filter_expressed = FALSE ){
  cell_types = colnames(binary_df)
  cell_geneslist = lapply(cell_types, function(cell){
    #gene_list = list()
    gene_list = row.names(binary_df[ binary_df[[cell]] != 0 , ])
    gene_list = gene_list[gene_list %in% expressed_genes]
    return(gene_list) })
  
  names(cell_geneslist) = cell_types  
  return(cell_geneslist) }


compute_median_expr=function(df,col_to_keep){
  df = df[,col_to_keep]
  df$genes = row.names(df)
  df_melt = reshape2::melt(df)
  return (median(df_melt$value)) }

compute_Avalue=function(expr_A, expr_B){ 
  return(0.5* ( log2(expr_A+1) + log2(expr_B+1) ) ) }
  #return(0.5* ( expr_A + expr_B ) ) }

compute_FC=function(expr_A, expr_B, median_expr){ 
  return(log2( (expr_A+median_expr ) / (expr_B+median_expr) )) }
  #return( (expr_A+median_expr ) / (expr_B+median_expr) ) } 

Get_value_RA = function(df, median_expr){
  A = colnames(df)[!str_detect(colnames(df), "vs|prop_expr")]
  B = colnames(df)[ str_detect(colnames(df), paste0("other_vs_.",A)) ]
  df[[paste0("FC_",A)]] = compute_FC( df[[A]] , df[[B]] ,median_expr)
  df[[paste0("Avalue_",A)]] = compute_Avalue( df[[A]] , df[[B]] )
  return(df)
}



```

# Processing of signature matrix {.tabset}

```{r signature_mat, fig.width=16, fig.height=8, eval=FALSE,warning=FALSE , message=FALSE , message=FALSE}

LM22 = read.csv("/home/acolajanni/Documents/work/doc/LM22/LM22_reduced.txt", sep='\t')
Sigmatrix = read.csv("/home/acolajanni/Documents/work/doc/LM22/Sigmatrix_Binary_reduced.txt", sep='\t')
Vallania = read.csv("/home/acolajanni/Documents/work/doc/LM22/Vallania_Binary_reduced.txt", sep='\t')
TIL10 = read.csv("/home/acolajanni/Documents/work/doc/LM22/TIL10_binary_2sd_reduced.txt", sep='\t')
features_expressed = colnames(read.csv("/home/acolajanni/Documents/work/data/Tabula_sapiens_immune_all/Blood_all_cell_expressed_genes.csv", 
                                       nrows = 1, header = TRUE, sep= '\t'))[2:32285]


LM22_genelist      = get_genes_from_binary(LM22, features_expressed, TRUE)
Sigmatrix_genelist = get_genes_from_binary(Sigmatrix, features_expressed, TRUE)
Vallania_genelist  = get_genes_from_binary(Vallania, features_expressed, TRUE)
TIL10_genelist     = get_genes_from_binary(TIL10, features_expressed, TRUE)


```

# Cell Types

```{r celltypes_og, fig.width=12, fig.height=9,warning=FALSE , message=FALSE}

cellTypes = read.table(file.path(data_dir,"16487_cell_types.txt"), header=FALSE, sep='\t')
label_count = cellTypes$V1
label_count = as.data.frame(table(label_count) )


ggplot( label_count, aes(x = reorder(label_count,Freq), y = Freq, fill=Freq ) )  +
  geom_col(position = position_dodge(0), width = 0.75) +
    geom_text(color="black",hjust=-0.2, size=3.8, angle=0,
            aes(y = ,label = Freq))+
  coord_flip(ylim = c(0,8000)  ) +
  labs(fill = "Geneset") +
  ylab('Frequency') + xlab('Cell type')+
  ggtitle(paste0("Frequency of cell types in Tabula Sapiens immune (Blood) dataset"))+
  scale_fill_viridis(discrete = FALSE, alpha=1, direction = -1, option='E', name="Absolute frequency") +
  theme_linedraw()+  
  theme( plot.title = element_text(size=15),
         axis.text.y = element_text(size = 12),
         axis.text.x = element_text(size = 12))

```

```{r celltypes, fig.width=15, fig.height=9,warning=FALSE , message=FALSE}

cellTypes = read.table(file.path(data_dir,"Celltypes_of_interest_3B.txt"), header=FALSE, sep='\t')
label_count = cellTypes$V1
label_count = as.data.frame(table(label_count) )

ggplot( label_count, aes(x = reorder(label_count,Freq), y = Freq, fill=Freq ) )  +
  geom_col(position = position_dodge(0), width = 0.75) +
    geom_text(color="black",hjust=-0.2, size=3.8, angle=0,
            aes(y = ,label = Freq))+
  coord_flip(ylim = c(0,8000)  ) +
  labs(fill = "Geneset") +
  ylab('Frequency') + xlab('Cell type')+
  ggtitle(paste0("Frequency of cell types in Tabula Sapiens immune (Blood) dataset"))+
  scale_fill_viridis(discrete = FALSE, alpha=1, direction = -1, option='E', name="Absolute frequency") +
  theme_linedraw()+  
  theme( plot.title = element_text(size=15),
         axis.text.y = element_text(size = 12),
         axis.text.x = element_text(size = 12))




n_cells = label_count
row.names(n_cells) = label_count$label_count
n_cells$label_count = NULL
n_cells = as.data.frame(t(n_cells))
n_cells$total = sum(n_cells[1,])
```

```{r celltypes_Liu, fig.width=15, fig.height=5,warning=FALSE , message=FALSE, include=FALSE}
data_dir_tmp = "/run/user/1001/gvfs/sftp:host=core.cluster.france-bioinformatique.fr,user=acolajanni/shared/projects/microbiome_translocation/data/scRNAseq/Liu_2021_cell/"

cellTypes = read.table(file.path(data_dir_tmp,"gene_set_84396_merged_immune_cell_labels.txt"), header=FALSE, sep='\t')
label_count = cellTypes$V1
label_count = as.data.frame(table(label_count) )


ggplot( label_count, aes(x = reorder(label_count,Freq), y = Freq, fill=Freq ) )  +
  geom_col(position = position_dodge(0), width = 0.75) +
    geom_text(color="black",hjust=-0.2, size=3.8, angle=0,
            aes(y = ,label = Freq))+
  coord_flip(ylim = c(0,40000)  ) +
  labs(fill = "Geneset") +
  ylab('Frequency') + xlab('Cell type')+
  ggtitle(paste0("Frequency of cell types in Liu et al. 2021, Cell. Healthy patients, Innate and Adaptative  "))+
  scale_fill_viridis(discrete = FALSE, alpha=1, direction = -1, option='E', name="Absolute frequency") +
  theme_linedraw()+  
  theme( plot.title = element_text(size=15),
         axis.text.y = element_text(size = 12),
         axis.text.x = element_text(size = 12))

```

# Filtering genes - Cell vs Cell {.tabset}

```{r LFC, fig.width=8, fig.height=6,warning=FALSE , message=FALSE}

## Retrieve merged Bcells data

groupwise_mean = read.table(file.path( data_dir,"8Celltype_wise_mean_all_cells.txt"), sep='\t', header = TRUE)
load(file.path(data_dir,"count_cell_prop.Rdata"))
groupwise_prop_full = count_cell_prop
count_cell_prop = NULL

# Compute median expr over all datatable
col = colnames(groupwise_mean)[!str_detect(colnames(groupwise_mean), "vs|genes" )]
expr_median = compute_median_expr(groupwise_mean,col )


colnames(groupwise_prop_full) = paste0(colnames(groupwise_prop_full),"_prop_expr")
groupwise_prop_full$genes = row.names(groupwise_prop_full)

groupwise_mean = merge(groupwise_mean, groupwise_prop_full, by='genes')


groupwise_B = groupwise_mean[,str_detect(colnames(groupwise_mean),"B")]
groupwise_B$genes = groupwise_mean$genes



################################################################################
# Retrieve other columns

groupwise_mean = read.table(file.path( data_dir,"Groupwise_mean_3B.txt"), sep='\t', header = TRUE)
load(file.path(data_dir,"count_cell_prop_3B.Rdata"))
groupwise_prop_full = count_cell_prop
count_cell_prop = NULL

# Compute median expr over all datatable
col = colnames(groupwise_mean)[!str_detect(colnames(groupwise_mean), "vs|genes" )]
expr_median = compute_median_expr(groupwise_mean,col )


colnames(groupwise_prop_full) = paste0(colnames(groupwise_prop_full),"_prop_expr")
groupwise_prop_full$genes = row.names(groupwise_prop_full)

groupwise_mean = merge(groupwise_mean, groupwise_prop_full, by='genes')

```

## CD4 vs CD8

**29 Features selected**

```{r LFC_CD84, fig.width=18, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_cd48 = groupwise_mean[str_detect(colnames(groupwise_mean),"CD4|CD8")]
FC_df_cd48$FC_cd48 = log2( (FC_df_cd48$CD4+expr_median ) / (FC_df_cd48$CD8+expr_median) )

elbow=kneedle_threshold(FC_df_cd48, FC_df_cd48$FC_cd48)
cd48 = Rprop_plot(FC_df_cd48, FC_x = "FC_cd48", y="CD4_prop_expr", 
                title = "Log Fold Change of CD4 vs CD8 depending on the proportion of CD4 cells expressing the genes", 
                #add_label = TRUE, 
                FC_threshold = 1, 
                prop_threshold = 2/n_cells$CD4, 
                prop_threshold_to_show = paste0("2 / ",as.character(n_cells$CD4)),
                highlight = "NCR1",
                FC_thresh2 = elbow$threshold) 

cd84 = Rprop_plot(FC_df_cd48, FC_x = "FC_cd48", y="CD8_prop_expr", 
                title = "Log Fold Change of CD4 vs CD8 depending on the proportion of CD8 cells expressing the genes", 
                #add_label = TRUE, 
                FC_threshold = 1, 
                prop_threshold = 2/n_cells$CD8,
                prop_threshold_to_show = paste0("2 / ",as.character(n_cells$CD8)),
                highlight = "NCR1",
                FC_thresh2 = elbow$threshold) 


cowplot::plot_grid(cd84$plot, cd48$plot, labels = 'AUTO', nrow = 1)



cd84_new = row.names(cd84$df[cd84$df$selection != "Not selected",])
cd48_new = row.names(cd48$df[cd48$df$selection != "Not selected",])


cd48_genes = unique(c(cd84_new,cd48_new))

#writeLines(cd48_genes, file.path(data_dir,'geneset_v2/4476_CD4vsCD8.txt') )
# writeLines(cd48_genes, file.path(git_geneset,'CD4vsCD8_4476.txt') )

```

### 4476 features

```{r cd48_barplot, fig.width=12, fig.height=10,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "CD4-CD8/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_cd48 = Permutation_list$`RF_permutation_First_CD4 vs CD8`
other_cd48$genes = row.names(other_cd48)

cd48_computation = importance_barplot(other_cd48, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 4476 genes selected in CD4 vs CD8 classification")
                                        
cd48_computation$plot

top24_cd48_genes = cd48_computation$ordered_permutation_dataframe[cd48_computation$ordered_permutation_dataframe$mean_importance >0,]$genes 

## writeLines(top24_cd48_genes, file.path(data_dir,'geneset_v2/24_CD4vsCD8.txt') )
# writeLines(top24_cd48_genes, file.path(git_geneset,'CD4vsCD8_24.txt') )

```

## NK vs CD8

**19 Features selected**

```{r LFC_NKCD8, fig.width=18, fig.height=10,  message=F, warning=F}
row.names(groupwise_mean)= groupwise_mean$genes
FC_df_NKcd8 = groupwise_mean[str_detect(colnames(groupwise_mean),"NK|CD8")]
FC_df_NKcd8$FC_NKCD8 = log2( (FC_df_NKcd8$NK_cell+expr_median ) / (FC_df_NKcd8$CD8+expr_median) )
FC_df_NKcd8$Avalue = 0.5* ( log2(FC_df_NKcd8$NK_cell+1) + log2(FC_df_NKcd8$CD8+1) )

elbow=kneedle_threshold(FC_df_NKcd8, FC_df_NKcd8$FC_NKCD8)
nkcd8 = Rprop_plot(FC_df_NKcd8, FC_x = "FC_NKCD8", y="NK_cell_prop_expr", 
                title = "Log Fold Change of NK vs CD8 depending on the proportion of NK cells expressing the genes", 
                #add_label = TRUE, 
                FC_threshold = 1, 
                prop_threshold = 2/n_cells$NK_cell,
                prop_threshold_to_show = paste0("2 / ",as.character(n_cells$NK_cell)),
                #highlight = test,
                FC_thresh2 = elbow$threshold) 

cd8nk = Rprop_plot(FC_df_NKcd8, FC_x = "FC_NKCD8", y="CD8_prop_expr", 
                title = "Log Fold Change of NK vs CD8 depending on the proportion of CD8 cells expressing the genes", 
                #add_label = TRUE, 
                FC_threshold = 1, 
                prop_threshold = 2/n_cells$CD8,
                prop_threshold_to_show = paste0("2 / ",as.character(n_cells$CD8)),
                #highlight = test,
                FC_thresh2 = elbow$threshold) 



cowplot::plot_grid(nkcd8$plot, cd8nk$plot, labels = 'AUTO', nrow = 1)

nkcd8_new = row.names(nkcd8$df[nkcd8$df$selection != "Not selected",])
cd8nk_new = row.names(cd8nk$df[cd8nk$df$selection != "Not selected",])

cd8nk_genes = unique(c(nkcd8_new,cd8nk_new))

## writeLines(cd8nk_genes, file.path(data_dir,'geneset_v2/1279_NKvsCD8.txt') )
# writeLines(cd8nk_genes, file.path(git_geneset,'NKvsCD8_1279.txt') )

```

### 1279 features

```{r nkcd8_barplot, fig.width=12, fig.height=7,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "NK-CD8/" ),
  pattern2= "/RF_permutation/Permutation_result_")


nkcd8 = Permutation_list$`RF_permutation_First_NK vs CD8`
nkcd8$genes = row.names(nkcd8)

nkcd8_computation = importance_barplot(nkcd8, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 1279 genes selected in NK vs CD8 classification")
                                        
nkcd8_computation$plot

top19_nkcd8_genes = nkcd8_computation$ordered_permutation_dataframe[nkcd8_computation$ordered_permutation_dataframe$mean_importance >0,]$genes 
## writeLines(top19_nkcd8_genes, file.path(data_dir,'geneset_v2/19_NKvsCD8.txt') )
# writeLines(top19_nkcd8_genes, file.path(git_geneset,'NKvsCD8_19.txt') )

```

## Naive vs Memory Bcells

**48 Features selected**

```{r LFC_naivememory_bcells, fig.width=18, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_naiveMem = groupwise_mean[str_detect(colnames(groupwise_mean),"naive|memory")]

FC_df_naiveMem$FC_nmB = log2( 
  (FC_df_naiveMem$memory_B_cell+expr_median ) / (FC_df_naiveMem$naive_B_cell+expr_median) )

elbow=kneedle_threshold(FC_df_naiveMem, FC_df_naiveMem$FC_nmB)

naivemem = Rprop_plot(FC_df_naiveMem, FC_x = "FC_nmB", y="naive_B_cell_prop_expr", 
                title = "Log Fold Change of Naive vs Memory Bcells depending on the proportion of naive Bcells expressing the genes", 
                #add_label = TRUE, 
                FC_threshold = 1, 
                prop_threshold = 2/n_cells$naive_B_cell,
                prop_threshold_to_show = paste0("2 / ",as.character(n_cells$naive_B_cell)),
                FC_thresh2 = elbow$threshold) 

memnaive = Rprop_plot(FC_df_naiveMem, FC_x = "FC_nmB", y="memory_B_cell_prop_expr", 
                title = "Log Fold Change of Naive vs Memory Bcells depending on the proportion of memory Bcells expressing the genes", 
                #add_label = TRUE, 
                FC_threshold = 1, 
                prop_threshold = 2/n_cells$memory_B_cell, 
                prop_threshold_to_show = paste0("2 / ",as.character(n_cells$memory_B_cell)),
                FC_thresh2 = elbow$threshold) 


cowplot::plot_grid(naivemem$plot, memnaive$plot, labels = 'AUTO', nrow = 1)


memnaive_new = row.names(memnaive$df[memnaive$df$selection != "Not selected",])
naivemem_new = row.names(naivemem$df[naivemem$df$selection != "Not selected",])
nmBcell_genes = unique(c(memnaive_new,naivemem_new))

## writeLines(nmBcell_genes, file.path(data_dir,'geneset_v2/2983_naiveVSmemory_B.txt') )
# writeLines(nmBcell_genes, file.path(git_geneset,'naiveVSmemory_B_2983.txt') )


```

### 2983 features

```{r B3_barplot, fig.width=12, fig.height=10,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Memory-Naive_Bcell/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_nm = Permutation_list$`RF_permutation_First_Naive vs Memory`
other_nm$genes = row.names(other_nm)

nm_computation = importance_barplot(other_nm, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 2983 genes selected in Memory vs Naive Bcells classification")
                                        
nm_computation$plot

top48_naive_mem = nm_computation$ordered_permutation_dataframe[nm_computation$ordered_permutation_dataframe$mean_importance >0,]$genes 

## writeLines(top48_naive_mem, file.path(data_dir,'geneset_v2/48_naiveVSmemory_B.txt') )
# writeLines(top48_naive_mem, file.path(git_geneset,'naiveVSmemory_B_48.txt') )

```


# Filtering genes - Cell vs all {.tabset}


## NK vs other

**247 Features selected**

```{r LFC_NK, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_NK = groupwise_mean[str_detect(colnames(groupwise_mean),"NK")]
FC_df_NK = Get_value_RA(FC_df_NK, expr_median)
FC_df_NK = FC_prop_expr(FC_df_NK)


nk = RA_plot(FC_df_NK, "Avalue_NK_cell", "FC_NK_cell", title = "RA plot NK vs other", add_label = FALSE)



elbow=kneedle_threshold(FC_df_NK, FC_df_NK$FC_NK_cell)
nk = Rprop_plot(FC_df_NK,FC_x = "FC_NK_cell", y="NK_cell_prop_expr", 
                title = "Log Fold Change of NK vs other depending on the proportion of NK cells expressing the genes", 
                highlight = "KLRB1",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 


nk$plot


nk_new = row.names(nk$df[nk$df$selection != "Not selected",])


## writeLines(nk_new, file.path(data_dir,'geneset_v2/2102_NK.txt') )
# writeLines(nk_new, file.path(git_geneset,'NK_2102.txt') )


```

### 2102 features

```{r nk_barplot, fig.width=12, fig.height=10,warning=FALSE , message=FALSE}
Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "NK/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_NK = Permutation_list$`RF_permutation_First_NK vs other`
other_NK$genes = row.names(other_NK)

NK_computation = importance_barplot(other_NK, 50, 
                                        filter = "zero", 
                                        title = "Mean feature importance of 50 permutations with 20102 genes selected in Other vs NK cell classification", 
                                    conf_interval = FALSE) 
                                        
NK_computation$plot + 
   theme(axis.text.y=element_blank(),
        axis.ticks.y=element_blank() 
        )

top568_nk = NK_computation$ordered_permutation_dataframe[NK_computation$ordered_permutation_dataframe$binf >0,]$genes 
## writeLines(top568_nk, file.path(data_dir,'geneset_v2/568_NK.txt') )
# writeLines(top568_nk, file.path(git_geneset,'NK_568.txt') )

```

### 568 features

```{r nk_barplot568, fig.width=12, fig.height=15,warning=FALSE , message=FALSE}
Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "NK/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_NK = Permutation_list$`RF_permutation_second_NK vs other`
other_NK$genes = row.names(other_NK)

NK_computation = importance_barplot(other_NK, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 20102 genes selected in Other vs NK cell classification")
                                        
NK_computation$plot + 
  theme(axis.text.y = element_text(size=5))

top182_nk = NK_computation$ordered_permutation_dataframe[NK_computation$ordered_permutation_dataframe$binf >0,]$genes 
top247_nk = NK_computation$ordered_permutation_dataframe[NK_computation$ordered_permutation_dataframe$mean_importance >0,]$genes 

## writeLines(top182_nk, file.path(data_dir,'geneset_v2/182_NK.txt') )
# writeLines(top247_nk, file.path(git_geneset,'NK_247.txt') )

```

## CD8 vs other

**116 Features selected**

```{r LFC_CD8, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_CD8 = groupwise_mean[str_detect(colnames(groupwise_mean),"CD8")]
FC_df_CD8 = Get_value_RA(FC_df_CD8, expr_median)
FC_df_CD8 = FC_prop_expr(FC_df_CD8)


elbow=kneedle_threshold(FC_df_CD8, FC_df_CD8$FC_CD8) 
cd8 = Rprop_plot(FC_df_CD8,FC_x = "FC_CD8", y="CD8_prop_expr", 
                title = "Log Fold Change of CD8 vs other depending on the proportion of CD8 cells expressing the genes",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

cd8$plot


cd8_new = row.names(cd8$df[cd8$df$selection != "Not selected",])
#test = FC_df_NK[row.names(FC_df_NK) %in% nkcd8_top170 , ]
## writeLines(cd8_new, file.path(data_dir,'geneset_v2/1877_CD8.txt') )
# writeLines(cd8_new, file.path(git_geneset,'CD8_1877.txt') )

```

### 1877 features

```{r cd8_barplot309, fig.width=12, fig.height=20,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "CD8/" ),
  pattern2= "/RF_permutation/Permutation_result_")


other_CD8 = Permutation_list$`RF_permutation_First_CD8 vs other`
other_CD8$genes = row.names(other_CD8)

cd8_computation = importance_barplot(other_CD8, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 1960 genes selected in Other vs CD8 classification")
                                        
cd8_computation$plot+ 
  theme(axis.text.y = element_text(size=5))


top309_cd8 = cd8_computation$ordered_permutation_dataframe[cd8_computation$ordered_permutation_dataframe$mean_importance >0,]$genes 
## writeLines(top309_cd8, file.path(data_dir,'geneset_v2/309_CD8.txt') )
# writeLines(top309_cd8, file.path(git_geneset,'CD8_309.txt') )

```

### 309 features

```{r cd8_barplot, fig.width=12, fig.height=18,warning=FALSE , message=FALSE}


other_CD8 = Permutation_list$`RF_permutation_second_CD8 vs other`
other_CD8$genes = row.names(other_CD8)

cd8_computation = importance_barplot(other_CD8, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 309 genes selected in Other vs CD8 classification")
                                        
cd8_computation$plot+ 
  theme(axis.text.y = element_text(size=10))


top116_cd8 = cd8_computation$ordered_permutation_dataframe[cd8_computation$ordered_permutation_dataframe$mean_importance >0,]$genes 
## writeLines(top116_cd8, file.path(data_dir,'geneset_v2/116_CD8.txt') )
# writeLines(top116_cd8, file.path(git_geneset,'CD8_116.txt') )

```

## CD4 vs other

**32 Features selected**

```{r LFC_CD4, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_CD4 = groupwise_mean[str_detect(colnames(groupwise_mean),"CD4")]
FC_df_CD4 = Get_value_RA(FC_df_CD4, expr_median)
FC_df_CD4 = FC_prop_expr(FC_df_CD4)


elbow=kneedle_threshold(FC_df_CD4, FC_df_CD4$FC_CD4) 
cd4 = Rprop_plot(FC_df_CD4,FC_x = "FC_CD4", y="CD4_prop_expr", 
                title = "Log Fold Change of CD4 vs other depending on the proportion of CD4 cells expressing the genes",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

cd4$plot


cd4_new = row.names(cd4$df[cd4$df$selection != "Not selected",])

## writeLines(cd4_new, file.path(data_dir,'geneset_v2/1375_CD4.txt') )
# writeLines(cd4_new, file.path(git_geneset,'CD4_1375.txt') )


```

### 1375 features

```{r cd4_barplot, fig.width=12, fig.height=14,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "CD4/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_CD4 = Permutation_list$`RF_permutation_First_CD4 vs other`
other_CD4$genes = row.names(other_CD4)

cd4_computation = importance_barplot(other_CD4, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 1572 genes selected in Other vs CD4 classification")
                                        
cd4_computation$plot+ 
  theme(axis.text.y = element_text(size=0))


CD4_top497 = cd4_computation$ordered_permutation_dataframe[cd4_computation$ordered_permutation_dataframe$mean_importance > 0,]$genes
## writeLines(CD4_top497, file.path(data_dir,'geneset_v2/497_CD4.txt') )
# writeLines(CD4_top497, file.path(git_geneset,'CD4_497.txt') )

```

### 497 features

```{r cd4_barplot497, fig.width=18, fig.height=12,warning=FALSE , message=FALSE}

other_CD4 = Permutation_list$`RF_permutation_second_CD4 vs other`
other_CD4$genes = row.names(other_CD4)

cd4_computation = importance_barplot(other_CD4, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 1572 genes selected in Other vs CD4 classification")
                                        
cd4_computation$plot+ 
  theme(axis.text.y = element_text(size=12))


CD4_top32 = cd4_computation$ordered_permutation_dataframe[cd4_computation$ordered_permutation_dataframe$mean_importance > 0,]$genes
## writeLines(CD4_top32, file.path(data_dir,'geneset_v2/32_CD4.txt') )
# writeLines(CD4_top32, file.path(git_geneset,'CD4_32.txt') )

```

## Bcells merged vs other

**Not include in the Signature matrix containing 630 features**

**5 Features selected**

```{r LFC_bcells, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_B)= groupwise_B$genes
FC_df_b = groupwise_B[str_detect(colnames(groupwise_B),"B_")]
FC_df_b = Get_value_RA(FC_df_b, expr_median)
FC_df_b = FC_prop_expr(FC_df_b)


elbow=kneedle_threshold(FC_df_b, FC_df_b$FC_B_cell) 
bcell = Rprop_plot(FC_df_b,FC_x = "FC_B_cell", y="B_cell_prop_expr", 
                title = "Log Fold Change of Bcells vs other depending on the proportion of Bcell cells expressing the genes",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

bcell$plot


bcell_new = row.names(bcell$df[bcell$df$selection != "Not selected",])



## writeLines(bcell_new, file.path(data_dir,'geneset_v2/1269_Bcell.txt') )

# writeLines(bcell_new, file.path(git_geneset,'Bcell_1269.txt') )

```

### 1269 features

```{r B_barplot, fig.width=12, fig.height=4,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Bcell/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_bcell = Permutation_list$`RF_permutation_First_Bcell vs other`
other_bcell$genes = row.names(other_bcell)


bcell_computation = importance_barplot(other_bcell, 50, 
                                        filter = "zero", 
                                        title = "Mean feature importance of 50 permutations with 1269 genes selected in Other vs Bcell classification")
                                        
bcell_computation$plot+ 
  theme(axis.text.y = element_text(size=10))



```

```{r B_barplot2, fig.width=16, fig.height=10,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Bcell/" ),
  pattern2= "/RF_permutation/",
  pattern1 = "importance_df")

other_bcell = Permutation_list$importance_df_RF_permutation_First.txt
other_bcell$genes = row.names(other_bcell)

other_bcell$X = NULL
other_bcell$X1 = 0
other_bcell = other_bcell[,c("X1","importance","genes")]
colnames(other_bcell)[2] = "mean_importance"

elbow=kneedle_threshold(other_bcell, other_bcell$mean_importance) 
elbow_low=kneedle_threshold(other_bcell, other_bcell$mean_importance, sens = 1000) 
elbow_low$plot

bcell_computation = importance_barplot(other_bcell[other_bcell$mean_importance >= elbow_low$threshold ,], 1, conf_interval = FALSE,
                                        filter = "zero", 
                                        title = "")
                                        
bcell_computation$plot + 
  theme(axis.text.y = element_text(size=5))+ 
  ggtitle("Random Forest Feature importance of the 129 most important genes with 1269 genes selected in Other vs Bcell classification (no permutation)",
          subtitle = "")



top129_Bcell = other_bcell[other_bcell$mean_importance >= elbow_low$threshold ,]$genes
# writeLines(top129_Bcell, file.path(data_dir,'geneset_v2/129_Bcell.txt') )
# writeLines(top129_Bcell, file.path(git_geneset,'Bcell_129.txt') )
```

### 129 features

```{r b_barplot129, fig.width=12, fig.height=10,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Bcell/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_B = Permutation_list$`RF_permutation_second_Bcell vs other`
other_B$genes = row.names(other_B)

B_computation = importance_barplot(other_B, 50, 
                                        filter = "zero", 
                                        title = "Mean feature importance of 50 permutations with 129 genes selected in Other vs Bcell classification")
                                        
B_computation$plot + 
  theme(axis.text.y = element_text(size=12))

top5_Bcell = B_computation$ordered_permutation_dataframe[B_computation$ordered_permutation_dataframe$mean_importance > 0 ,]$genes

```

## Naive Bcell vs other

**4 Features selected**

```{r LFC_naive_bcells, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_naiveB = groupwise_mean[str_detect(colnames(groupwise_mean),"naive")]
FC_df_naiveB = Get_value_RA(FC_df_naiveB, expr_median)
FC_df_naiveB = FC_prop_expr(FC_df_naiveB)


elbow=kneedle_threshold(FC_df_naiveB, FC_df_naiveB$naive_B_cell) 
naiveB = Rprop_plot(FC_df_naiveB,FC_x = "FC_naive_B_cell", y="naive_B_cell_prop_expr", 
                title = "Log Fold Change of Naive B cell vs other depending on the proportion of CD8 cells expressing the genes",
                highlight = "NCR1",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

naiveB$plot


naiveB_genes = row.names(naiveB$df[naiveB$df$selection != "Not selected",])
#writeLines(naiveB_genes, file.path(data_dir,'geneset_v2/4935_naiveB.txt') )
# writeLines(naiveB_genes, file.path(git_geneset,'naiveB_4935.txt') )

```

### 4935 features

```{r naiveB_barplot, fig.width=12, fig.height=10,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Naive_Bcell/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_naiveB = Permutation_list$`RF_permutation_First_naive_Bcell vs other`
other_naiveB$genes = row.names(other_naiveB)

naiveB_computation = importance_barplot(other_naiveB, 50, 
                                        filter = "zero", 
                                        title = "Mean feature importance of 50 permutations with 4935 genes selected in Other vs naive Bcell classification",
                                        conf_interval = FALSE)
                                        
naiveB_computation$plot + 
  theme(axis.text.y = element_text(size=0))

#other_naiveB[other_naiveB$mean_importance < 0 ,]

```

```{r naiveB_barplot2, fig.width=12, fig.height=8,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Naive_Bcell/" ),
  pattern2= "/RF_permutation/",
  pattern1 = "importance_df")

other_nbcell = Permutation_list$importance_df_RF_permutation_First.txt
other_nbcell$genes = row.names(other_nbcell)

other_nbcell$X = NULL
other_nbcell$X1 = 0
other_nbcell = other_nbcell[,c("X1","importance","genes")]
colnames(other_nbcell)[2] = "mean_importance"

elbow=kneedle_threshold(other_nbcell, other_nbcell$mean_importance, sens = 50) 
elbow$plot

elbow1=kneedle_threshold(other_nbcell, other_nbcell$mean_importance, sens = 1) 




nbcell_computation = importance_barplot(other_nbcell[other_nbcell$mean_importance >= elbow$threshold ,], 
                                       #other_nbcell
                                       1, conf_interval = FALSE,
                                        filter = "zero", 
                                        title = "Random Forest Feature importance of the 281 most important genes with 4935 genes selected in Other vs Bcell classification") 
  
                                        
nbcell_computation$plot + 
  theme(axis.text.y = element_text(size=0))

top281_nbcell = other_nbcell[other_nbcell$mean_importance >= elbow$threshold ,]$genes

## writeLines(top281_nbcell, file.path(data_dir,'geneset_v2/281_naiveB.txt') )
# writeLines(naiveB_genes, file.path(git_geneset,'naiveB_281.txt') )

```

### 281 features

```{r niaveB_barplot281, fig.width=16, fig.height=6,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Naive_Bcell/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_naiveB = Permutation_list$`RF_permutation_second_naive_Bcell vs other`
other_naiveB$genes = row.names(other_naiveB)

naiveB_computation = importance_barplot(other_naiveB, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 4935 genes selected in Other vs naive Bcell classification")
                                        
naiveB_computation$plot + 
  theme(axis.text.y = element_text(size=14))

top4_naiveB = other_naiveB[other_naiveB$mean_importance > 0 ,]$genes
# writeLines(top4_naiveB, file.path(git_geneset,'naiveB_4.txt') )

```

## Memory Bcell vs other

**32 Features selected**

```{r LFC_mem_bcells, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_memB = groupwise_mean[str_detect(colnames(groupwise_mean),"memory")]
FC_df_memB = Get_value_RA(FC_df_memB, expr_median)
FC_df_memB = FC_prop_expr(FC_df_memB)


elbow=kneedle_threshold(FC_df_memB, FC_df_memB$memory_B_cell) 
memB = Rprop_plot(FC_df_memB,FC_x = "FC_memory_B_cell", y="memory_B_cell_prop_expr", 
                title = "Log Fold Change of Memory B cell vs other depending on the proportion of CD8 cells expressing the genes",
                highlight = "NCR1",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

memB$plot

memB_genes = row.names(memB$df[memB$df$selection != "Not selected",])
## writeLines(memB_genes, file.path(data_dir,'geneset_v2/6855_memoryB.txt') )
# writeLines(memB_genes, file.path(git_geneset,'memoryB_6855.txt') )

```

### 6855 features

```{r memB_barplot, fig.width=12, fig.height=10,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Memory_Bcell/" ),
  pattern2= "/RF_permutation/Permutation_result_")

# Attendre la fin des calculs

other_memB = Permutation_list$`RF_permutation_missing_memory_Bcell vs other`
other_memB$genes = row.names(other_memB)

memB_computation = importance_barplot(other_memB, 50, 
                                        filter = "zero", 
                                        title = "Mean feature importance of 50 permutations with 4935 genes selected in Other vs nnaive Bcell classification")
                                        
memB_computation$plot + 
  theme(axis.text.y = element_text(size=9))

top33_memB = memB_computation$ordered_permutation_dataframe[memB_computation$ordered_permutation_dataframe$mean_importance > 0 ,]$genes
# writeLines(top33_memB, file.path(git_geneset,'memoryB_33.txt') )

```

## Neutrophils vs other

**5 Features selected**

```{r LFC_neutro, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_neutro = groupwise_mean[str_detect(colnames(groupwise_mean),"Neutro")]
FC_df_neutro = Get_value_RA(FC_df_neutro, expr_median)
FC_df_neutro = FC_prop_expr(FC_df_neutro)


elbow=kneedle_threshold(FC_df_neutro, FC_df_neutro$FC_Neutrophil) 
neutro = Rprop_plot(FC_df_neutro,FC_x = "FC_Neutrophil", y="Neutrophil_prop_expr", 
                title = "Log Fold Change of Neutrophil vs other depending on the proportion of Neutrophil cells expressing the genes",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

neutro$plot


neutro_new = row.names(neutro$df[neutro$df$selection != "Not selected",])
## writeLines(neutro_new, file.path(data_dir,'geneset_v2/984_Neutro.txt') )
# writeLines(neutro_new, file.path(git_geneset,'Neutro_984.txt') )


```

### 984 features

```{r neutro_barplot, fig.width=12, fig.height=6,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Neutrophil/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_neutro = Permutation_list$`RF_permutation_First_Neutrophil vs other`
other_neutro$genes = row.names(other_neutro)

neutro_computation = importance_barplot(other_neutro, 50, 
                                        filter = "zero", 
                                        title = "Mean feature importance of 50 permutations with 984 genes selected in Other vs neutrophil classification")
                                        
neutro_computation$plot + 
  theme(axis.text.y = element_text(size=10))

neutro_top6 = neutro_computation$ordered_permutation_dataframe[neutro_computation$ordered_permutation_dataframe$mean_importance>0,]$genes
## writeLines(neutro_top6, file.path(data_dir,'geneset_v2/6_Neutro.txt') )
# writeLines(neutro_top6, file.path(git_geneset,'Neutro_6.txt') )

```

## Plasmocyte vs other

**32 Features selected**

```{r LFC_plasmo, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_Plasmo = groupwise_mean[str_detect(colnames(groupwise_mean),"Plasmo")]
FC_df_Plasmo = Get_value_RA(FC_df_Plasmo, expr_median)
FC_df_Plasmo = FC_prop_expr(FC_df_Plasmo)


elbow=kneedle_threshold(FC_df_Plasmo, FC_df_Plasmo$FC_Plasmocyte) 
plasmo = Rprop_plot(FC_df_Plasmo,FC_x = "FC_Plasmocyte", y="Plasmocyte_prop_expr", 
                title = "Log Fold Change of Plasmocyte vs other depending on the proportion of Plasmocyte cells expressing the genes",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

plasmo$plot


plasmo_new = row.names(plasmo$df[plasmo$df$selection != "Not selected",])
## writeLines(plasmo_new, file.path(data_dir,'geneset_v2/4443_Plasmocyte.txt') )
# writeLines(plasmo_new, file.path(git_geneset,'Plasmocyte_4443.txt') )


```

### 4443 features

```{r plasmo_barplot, fig.width=12, fig.height=10,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Plasmocyte/" ),
  pattern2= "/RF_permutation/Permutation_result_")



other_plasmo = Permutation_list$`RF_permutation_First_Plasmocyte vs other`
other_plasmo$genes = row.names(other_plasmo)

plasmo_computation = importance_barplot(other_plasmo, 50, 
                                        filter = "zero", 
                                        title = "Mean feature importance of 50 permutations with 4443 genes selected in Other vs Plasmocyte classification")
                                        
plasmo_computation$plot + 
  theme(axis.text.y = element_text(size=10))

plasmo_top32 = plasmo_computation$ordered_permutation_dataframe[plasmo_computation$ordered_permutation_dataframe$mean_importance>0,]$genes
## writeLines(plasmo_top32, file.path(data_dir,'geneset_v2/32_Plasmocyte.txt') )
# writeLines(plasmo_top32, file.path(git_geneset,'Plasmocyte_32.txt') )

```

## Monocyte vs other

**144 Features selected**

```{r LFC_mono, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_mono = groupwise_mean[str_detect(colnames(groupwise_mean),"Mono")]
FC_df_mono = Get_value_RA(FC_df_mono, expr_median)
FC_df_mono = FC_prop_expr(FC_df_mono)


elbow=kneedle_threshold(FC_df_mono, FC_df_mono$FC_Monocyte) 
mono = Rprop_plot(FC_df_mono,FC_x = "FC_Monocyte", y="Monocyte_prop_expr", 
                title = "Log Fold Change of Monocyte vs other depending on the proportion of Monocyte cells expressing the genes",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

mono$plot


mono_new = row.names(mono$df[mono$df$selection != "Not selected",])
## writeLines(mono_new, file.path(data_dir,'geneset_v2/3182_Monocyte.txt') )
# writeLines(mono_new, file.path(git_geneset,'Monocyte_3182.txt') )


```

### 3182 features

```{r mono_barplot, fig.width=12, fig.height=14,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Monocyte/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_mono = Permutation_list$`RF_permutation_First_Monocyte vs other` 
other_mono$genes = row.names(other_mono)

mono_computation = importance_barplot(other_mono, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 3182 genes selected in Other vs Monocyte classification")
                                        
mono_computation$plot + 
  theme(axis.text.y = element_text(size=7))


mono_top97 = mono_computation$ordered_permutation_dataframe[mono_computation$ordered_permutation_dataframe$binf > 0,]$genes
mono_top144 = mono_computation$ordered_permutation_dataframe[mono_computation$ordered_permutation_dataframe$mean_importance > 0,]$genes
# writeLines(mono_top144, file.path(git_geneset,'Monocyte_144.txt') )

```

# Combining Features

```{r combine_features3.2, fig.width=16, fig.height=10,warning=FALSE , message=FALSE}

geneset_list_v2 = list(
  #"NK"        = unique(c(top182_nk,top19_nkcd8_genes)),
  "NK"        = unique(c(top247_nk,top19_nkcd8_genes)),
  "Bcell"     = top5_Bcell,#top129_Bcell,

  "Neutrophil"= neutro_top6,
  "Monocyte"  = mono_top144,#mono_top97,
  "Plasmocyte"= plasmo_top32,
  
  "CD4"       = unique(c(top24_cd48_genes, CD4_top32)),
  "CD8"       = unique(c(top24_cd48_genes, top116_cd8, top19_nkcd8_genes)), 
  
  "MemoryB"   = unique(c(top48_naive_mem, top33_memB)), 
  "NaiveB"    = unique(c(top4_naiveB,top48_naive_mem))
  )
  
Full_geneset_list_v2 = unique(unlist(geneset_list_v2))


upset(fromList(geneset_list_v2),     
      sets.bar.color = "#56B4E9", order.by = "degree", nsets = 9,
      empty.intersections = NULL, set_size.show = TRUE, mb.ratio = c(0.7, 0.3),
      text.scale = 1.75, set_size.scale_max = 300)


upset(fromList(geneset_list_v2[c("NaiveB","MemoryB","Bcell")]),     
      sets.bar.color = "#56B4E9", order.by = "degree", nsets = 9,
      empty.intersections = NULL, set_size.show = TRUE, mb.ratio = c(0.7, 0.3),
      text.scale = 1.75, set_size.scale_max = 90)

```

## Comparing with the previous signature matrix (518 genes / 7 celltypes)

```{r combine_features3.3, fig.width=16, fig.height=10,warning=FALSE , message=FALSE}
common_cells = geneset_list_v2
common_cells$NaiveB = NULL
common_cells$MemoryB = NULL
common_cells_flat = unique(unlist(common_cells))


load(file = "/home/acolajanni/Documents/work/doc/LM22/Geneset_7types_TabulaSapiens.Rdata")

Full_geneset_7types = unique(unlist(geneset_list_7types))


upset(fromList(list("518 genelist" = Full_geneset_7types, "new list" = common_cells_flat)),     
      sets.bar.color = "#56B4E9", order.by = "degree", nsets = 9,
      empty.intersections = NULL, set_size.show = TRUE, mb.ratio = c(0.7, 0.3),
      text.scale = 1.75, set_size.scale_max = 700)

plot_list = list()
for (cell in names(geneset_list_7types)) {
  
  cell_list = list(
    geneset_list_7types[[cell]] , 
    geneset_list_v2[[cell]] )
  
  names(cell_list) = c(paste0("old ",cell),paste0("new ",cell))
  
  plot_list[[cell]]=ggvenn::ggvenn(cell_list ) 


} 

print(plot_list)
```

```{r combine_features3.4, fig.width=16, fig.height=10, include=FALSE}

binary_1B = geneset_to_binary_df(common_cells)

geneset_list_v2_noB = geneset_list_v2
geneset_list_v2_noB$Bcell = NULL
binary_3B = geneset_to_binary_df(geneset_list_v2_noB)




################# 
# Geneset Bcell


# writeLines(common_cells_flat, file.path(path,"/results/Tabula_sapiens/LM22/signature_genes_1B.txt"))
# 
# write.table(binary_1B, file.path(path,"/results/Tabula_sapiens/LM22/binary_df_1B.txt"),
#              quote = FALSE, row.names = TRUE, sep='\t')
            
            
#################
# Geneset Bmemory / B naive

geneset_3B = geneset_list_v2
geneset_3B$Bcell = NULL
geneset_3B_flat = unique(unlist(geneset_3B)) 
binary_3B = geneset_to_binary_df(geneset_3B)  
  
# writeLines(geneset_3B_flat, file.path(path,"/results/Tabula_sapiens/LM22/signature_genes_3B.txt"))

# write.table(binary_3B, file.path(path,"/results/Tabula_sapiens/LM22/binary_df_3B.txt"),
#             quote = FALSE, row.names = TRUE, sep='\t')

# write.table(binary_3B, file.path(git_dir,"/Signature_matrix/Signature_TabulaSapiens_630_8types_Binary.txt"),
#              quote = FALSE, row.names = TRUE, sep='\t')

#################
# Geneset Kalidou : B  / CD4 / CD8 / Neutrophil / 

geneset_5types = geneset_list_v2
geneset_5types$Monocyte = NULL
geneset_5types$Plasmocyte = NULL
geneset_5types$MemoryB = NULL
geneset_5types$NaiveB = NULL

geneset_5t_flat = unique(unlist(geneset_5types)) 
binary_5t = geneset_to_binary_df(geneset_5types)  
  
# writeLines(geneset_5t_flat, file.path(path,"/results/Tabula_sapiens/LM22/signature_genes_5types_v2.txt"))
# 
# write.table(binary_5t, file.path(path,"/results/Tabula_sapiens/LM22/binary_df_5types_v2.txt"),
#             quote = FALSE, row.names = TRUE, sep='\t')



```

# Mean Expression heatmap

```{r mean_expr_heat630, fig.width=16, fig.height=9, include = FALSE}
geneset_new = unique(unlist(geneset_list_v2))

groupwise_test = merge(groupwise_mean, groupwise_B, by="genes")

heatmap = groupwise_test[groupwise_test$genes %in% geneset_new,]
heatmap = heatmap[,! str_detect(colnames(heatmap), "other|Erythrocyte|prop_expr")]


row.names(heatmap) = heatmap$genes
heatmap$genes = NULL

# write.table(heatmap, file.path(data_dir,"/geneset_8types/525_signature_TabulaSapiens.txt"),
#          quote = FALSE, row.names = TRUE, sep='\t')


genes_annotation = heatmap
genes_annotation$NK_cell = ifelse(row.names(genes_annotation) %in% geneset_list_v2$NK, "yes" , "no")
genes_annotation$CD4 = ifelse(row.names(genes_annotation) %in% geneset_list_v2$CD4, "yes" , "no")
genes_annotation$CD8 = ifelse(row.names(genes_annotation) %in% geneset_list_v2$CD8,"yes" , "no")
genes_annotation$Neutrophil = ifelse(row.names(genes_annotation) %in% geneset_list_v2$Neutrophil,"yes" , "no")
genes_annotation$B_cell = ifelse(row.names(genes_annotation) %in% geneset_list_v2$Bcell,"yes" , "no")
genes_annotation$naive_B_cell = ifelse(row.names(genes_annotation) %in% geneset_list_v2$NaiveB,"yes" , "no")
genes_annotation$memory_B_cell = ifelse(row.names(genes_annotation) %in% geneset_list_v2$MemoryB,"yes" , "no")
genes_annotation$Plasmocyte = ifelse(row.names(genes_annotation) %in% geneset_list_v2$Plasmocyte,"yes" , "no")
genes_annotation$Monocyte = ifelse(row.names(genes_annotation) %in% geneset_list_v2$Monocyte,"yes" , "no")


color_yes = "red"
#color_no = "darkgrey"
color_no = "#666666"

my_colour = list(
  B_cell = c(yes = color_yes, no = color_no),
  naive_B_cell = c(yes = color_yes, no = color_no),
  memory_B_cell = c(yes = color_yes, no = color_no),
  Neutrophil = c(yes = color_yes, no = color_no),
  CD4 = c(yes = color_yes, no = color_no), 
  NK_cell = c(yes = color_yes, no = color_no) ,
  CD8 = c(yes = color_yes, no = color_no),
  Plasmocyte = c(yes = color_yes, no = color_no),
  Monocyte = c(yes = color_yes, no = color_no)
    )


pheatmap(t(heatmap), scale = "column", 
         cutree_cols = 8,
         angle_col = "45", main = "Mean expression heatmap of the XXX features used to predict cell types V2",
         clustering_distance_rows = 'correlation', 
         clustering_distance_cols = 'correlation',
         clustering_method = "ward.D2",
         annotation_col = genes_annotation, annotation_colors = my_colour,
         color = colorRampPalette(c("navy", "#DDDDDD", "firebrick"))(11),
         show_colnames = FALSE)

```

```{r mean_expr_heatnoB, fig.width=16, fig.height=9,warning=FALSE , message=FALSE}
geneset_new = unique(unlist(geneset_list_v2_noB))
groupwise_test = merge(groupwise_mean, groupwise_B, by="genes")

heatmap = groupwise_mean[groupwise_mean$genes %in% geneset_new,]
heatmap = heatmap[,! str_detect(colnames(heatmap), "other|Erythrocyte|prop_expr")]

row.names(heatmap) = heatmap$genes
heatmap$genes = NULL

genes_annotation$B_cell = NULL
my_colour$B_cell = NULL

#row.names(heatmap) = NULL

setHook("grid.newpage", function() pushViewport(viewport(x=1,y=1,width=1, height=0.95, name="vp", just=c("right","top"))), action="prepend")
pheatmap(t(heatmap), scale = "column", 
         cutree_cols = 10,
         angle_col = "45", main = "Mean expression heatmap of the 630 features used to predict 8 cell types",
         clustering_distance_rows = 'correlation', 
         clustering_distance_cols = 'correlation',
         clustering_method = "ward.D2",
         annotation_col = genes_annotation, annotation_colors = my_colour,
         color = colorRampPalette(c("navy", "#DDDDDD", "firebrick"))(11),
         show_colnames = FALSE)
         #col_title = "Genes"
         
setHook("grid.newpage", NULL, "replace")
grid.text("Genes", y=-0.02, gp=gpar(fontsize=16))
```

```{r mean_, fig.width=16, fig.height=9, include=FALSE ,warning=FALSE , message=FALSE}


warning=FALSE

### B memory + B naive

geneset_new = unique(unlist(geneset_list_v2_noB))
#save(TabulaSapiens_geneset, file = file.path(git_geneset,"TabulaSapiens_geneset.Rdata"))

#groupwise_test = merge(groupwise_mean, groupwise_B, by="genes")

heatmap = groupwise_mean[groupwise_mean$genes %in% geneset_new,]
heatmap = heatmap[,! str_detect(colnames(heatmap), "other|Erythrocyte|prop_expr")]


row.names(heatmap) = heatmap$genes
heatmap$genes = NULL

# write.table(heatmap, file.path(data_dir,"/geneset_v2/Signature_TabulaSapiens_586_8types_v2.txt"),
# quote = FALSE, row.names = TRUE, sep='\t')
# write.table(heatmap, file.path(path,"results/Tabula_sapiens/LM22/Signature_TabulaSapiens_586_8types_v2.txt"),
# quote = FALSE, row.names = TRUE, sep='\t')

# write.table(heatmap, file.path(data_dir,"/geneset_v2/Signature_TabulaSapiens_630_8types_v2.txt"),
# quote = FALSE, row.names = TRUE, sep='\t')
# 
# write.table(heatmap, file.path(path,"results/Tabula_sapiens/LM22/Signature_TabulaSapiens_630_8types_v2.txt"),
# quote = FALSE, row.names = TRUE, sep='\t')
# 
# write.table(heatmap, file.path(git_dir,"Signature_matrix/Signature_TabulaSapiens_630_8types.txt"),
# quote = FALSE, row.names = TRUE, sep='\t')

### B (no naive/memory)

geneset_new = unique(unlist(common_cells))

groupwise_test = merge(groupwise_mean, groupwise_B, by="genes")

heatmap = groupwise_test[groupwise_test$genes %in% geneset_new,]
heatmap = heatmap[,! str_detect(colnames(heatmap), "other|Erythrocyte|prop_expr|aive|emory")]


row.names(heatmap) = heatmap$genes
heatmap$genes = NULL

# write.table(heatmap, file.path(data_dir,"/geneset_v2/Signature_TabulaSapiens_557_Bcell_unique.txt"),
#         quote = FALSE, row.names = TRUE, sep='\t')
# write.table(heatmap, file.path(path,"results/Tabula_sapiens/LM22/Signature_TabulaSapiens_557_Bcell_unique.txt"),
#         quote = FALSE, row.names = TRUE, sep='\t')

### 5 types

geneset_new = unique(unlist(geneset_5types))

groupwise_test = merge(groupwise_mean, groupwise_B, by="genes")

heatmap = groupwise_test[groupwise_test$genes %in% geneset_new,]
heatmap = heatmap[,! str_detect(colnames(heatmap), "other|Erythrocyte|prop_expr|aive|emory")]


row.names(heatmap) = heatmap$genes
heatmap$genes = NULL

# # write.table(heatmap, file.path(data_dir,"/geneset_v2/Signature_TabulaSapiens_445_5types.txt"),
#         quote = FALSE, row.names = TRUE, sep='\t')
# # write.table(heatmap, file.path(path,"results/Tabula_sapiens/LM22/Signature_TabulaSapiens_445_5types.txt"),
#         quote = FALSE, row.names = TRUE, sep='\t')

```
