---
title: "Feature selection of cell types in scRNAseq Tabula sapiens (50k Blood cells) "
author: "Antonin Colajanni"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: yes
    toc_depth: 2
    toc_float: 
      collapsed: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

# Tabula Sapiens Analyses : RF feature reduction

```{r  fix_knit, fig.width=12, fig.height=9, eval=TRUE, include=FALSE,warning=FALSE , message=FALSE , message=FALSE}
library(rmarkdown)

# setwd("/home/acolajanni/Documents/work/results/Tabula_sapiens/markdown/")
# clean_tmpfiles_mod <- function() {
#   message("Calling clean_tmpfiles_mod()")
# }
# 
# assignInNamespace("clean_tmpfiles", clean_tmpfiles_mod, ns = "rmarkdown")

library(purrr)
rate <- rate_backoff(pause_base = 0.1, pause_min = 0.005, max_times = 10)
insistent_render <- insistently(rmarkdown::render, rate, quiet = FALSE)


```

```{r import, echo=T, results='hide', message=F, warning=F, eval=T}

# I M P O R T
library(stringr)
library (plyr) 
library(ggplot2)
library(viridis)
library(UpSetR)
library(ggrepel)
library(factoextra)
library(FactoMineR)
library(viridis)
library(RColorBrewer)
library(dplyr)
library(pheatmap)
library(DT) 
library(cluster)
library(grid)
# V A R I A B L E S
#main.dir = "/run/user/1001/gvfs/sftp:host=curta.mcia.fr,user=acolajanni/gpfs/home/acolajanni"
path = "/run/user/1001/gvfs/sftp:host=core.cluster.france-bioinformatique.fr,user=acolajanni/shared/projects/microbiome_translocation"
local_path = "/home/acolajanni/Documents/work/"
#setwd(main.dir)

#path = "/home/acolajanni/Documents/work/"
data_dir = file.path(path, "data/Tabula_sapiens_immune_all")

result_dir = file.path(path, "results/Tabula_sapiens/RF_filter_50k/")
#setwd(path)

git_dir = file.path(local_path,"GITHUB/Signature_matrix_TabulaSapiens/")
git_geneset = file.path(git_dir,"genesets/")
################################################################################
# F U N C T I O N S

create_dir = function(directory){
  if ( ! file.exists(directory)){
    dir.create(directory)  }
  return(directory) }

replace_label = function(label, name_mapping){
  
  label_names = names(name_mapping)
  label = ifelse(str_detect(label, label_names[1]),name_mapping[1],label)
  label = ifelse(str_detect(label,label_names[2]),name_mapping[2],label)
  label = ifelse(label %in% c(name_mapping[1], name_mapping[2]), yes = label, no = name_mapping[3])
  #label=as.numeric(label)
  return(label) }


confidence_interval = function(perm_df, n_perm = 20) {
  t.score=qt(p=0.025, df=n_perm-1 ,lower.tail = FALSE)
  perm_df$se = perm_df$std_importance/sqrt(n_perm-1)
  perm_df$margin_error = t.score*perm_df$se
  perm_df$binf = perm_df$mean_importance - perm_df$margin_error
  perm_df$bsup = perm_df$mean_importance + perm_df$margin_error
  
  return(perm_df) }


importance_barplot = function(permutation_df,permutation_number=50, filter="mean", title="Mean feature importance barplot",
                              conf_interval = TRUE){
  if (conf_interval){
    permutation_df = confidence_interval(permutation_df, permutation_number) 
  }
  else {
    permutation_df$binf = permutation_df$mean_importance
    permutation_df$bsup = permutation_df$mean_importance
  }
  non_null_feature = length(permutation_df[permutation_df$mean_importance != 0 , ]$mean_importance)
  pos_feature = length(permutation_df[permutation_df$mean_importance > 0 , ]$mean_importance)

  
  if (filter == "mean") { 
    permutation_df = permutation_df[permutation_df$mean_importance > 0 , ]
    subtitle = paste(non_null_feature, "Genes with non null feature importance among which", pos_feature, "have a positive importance") }
  else {  
    sig_pos_feature = length(permutation_df[permutation_df$binf > 0 , ]$mean_importance)
    subtitle = paste(non_null_feature, "Genes with non null feature importance among which", pos_feature, "have a positive importance and \n ",
                     sig_pos_feature, "are significantly different than zero.")
    
    if (filter == "bsup") { permutation_df = permutation_df[permutation_df$bsup > 0 , ] }
    else if (filter == "binf") { permutation_df = permutation_df[permutation_df$binf > 0 , ] }
    else if (filter == "zero") { permutation_df = permutation_df[permutation_df$mean_importance != 0 , ] }
  }
  
  
  permutation_df$genes = row.names(permutation_df)
  n_gene = length(row.names(permutation_df))
  ordered_permutation_df = permutation_df[order(permutation_df$mean_importance, decreasing = TRUE),]
  
  plot = ggplot( permutation_df, aes(x = reorder(genes,mean_importance), y = mean_importance, fill=mean_importance ) )  +
    geom_col(position = position_dodge(0), width = 0.75) + 
    geom_errorbar(aes(ymin=binf, ymax=bsup ),
                size=.3,width=.2, position=position_dodge(.9)) +
    geom_hline(aes(yintercept = 0), alpha=0.20) +
    coord_flip() +
    labs(fill = "Cluster") +
    ylab('Mean Permutation Feature Importance') + xlab('Gene name')+
    ggtitle(title, subtitle = subtitle ) +
  
    scale_fill_viridis(discrete = FALSE, alpha=0.65, direction = -1, option='C', name="Mean Feature Importance") +
    theme_linedraw()+  
    theme( plot.title = element_text(size=15),
           axis.text.y = element_text(size = 8))
  return(list("plot"=plot, 
              "ordered_permutation_dataframe" = ordered_permutation_df[,-c(1:permutation_number)],
              "permutation_dataframe"= permutation_df)) }


add_annotation_line_barplot_importance = function(plot, permutation_df, last_gene, annotation = "most important features", adjust_annotation = .25, adjust_annotationX=0.85){
  gene_value = permutation_df[permutation_df$genes == last_gene,]$mean_importance
  gene_rank = length(permutation_df[permutation_df$mean_importance >= gene_value,]$genes)
  
  plot = plot +
    geom_vline( xintercept = which(permutation_df$genes == last_gene)-0.5, linetype='dashed' )+
    annotate("text", x=which(permutation_df$genes == last_gene)+adjust_annotation , y=max(permutation_df$mean_importance)*adjust_annotationX,
             label= paste(gene_rank,annotation) )
    return(plot) }

RA_plot = function(FC_df, x, y, title = "RA plot", highlight=c('TOP20'), annotation = "Most Important Features", vjust = 2, force=3, y_scale = 0, add_label=TRUE){
  
  FC_df$selection = ifelse(FC_df[[x]] > quantile(FC_df[[x]],0.75) & abs(FC_df[[y]]) > 1 , 
                            yes = "Selected",no = "Not selected")
  selected_number= length(FC_df$selection[FC_df$selection == "Selected"])
    
  if (highlight[1] == "TOP20") {
    FC_df$selection = ifelse(abs(FC_df[[y]]) > sort(abs(FC_df[[y]]),TRUE )[21]  , 
                            yes = "20 Highest FC",no = FC_df$selection)
    FC_df$label = ifelse(FC_df$selection == "20 Highest FC", 
                          yes=row.names(FC_df), no="") }
  else{
    FC_df$selection = ifelse(row.names(FC_df) %in% highlight , 
                            yes = annotation, no = FC_df$selection)
    FC_df$label = ifelse(FC_df$selection == annotation, 
                          yes=row.names(FC_df), no="") }
  
  
  maximum_y = max(abs(FC_df[[y]])) + y_scale
  title =paste( c(title, "\n Gene selected :",selected_number),collapse = " " )
  
  plot = ggplot(FC_df, aes(x=FC_df[[x]], y=FC_df[[y]]))+
    geom_point(aes(color=selection, alpha=selection, size = selection))+
    geom_smooth(color="darkred",size=0.5)+
    annotate("text", x=2.4, y=-1-0.5, label="log2 FC = -1", colour = "darkblue") +
    geom_hline(yintercept = -1, color='darkblue',linetype='dashed') +
    annotate("text", x=2.4, y= 1+0.5, label="log2 FC = 1", colour = "darkblue") +
    coord_cartesian(ylim=c(-maximum_y,maximum_y))+
    geom_hline(yintercept = 1, color='darkblue',linetype='dashed') +
    geom_vline(xintercept = quantile(FC_df[[x]],0.75),linetype='dashed') +
    annotate("text", y=-5.5, x= quantile(FC_df[[x]],0.75)/4, 
           label="3rd quartile \n of mean expression", colour = "black") +
    
    scale_color_manual(values = c("orangered","#777777","#00539CFF"))+
    scale_alpha_manual(values=c(1,0.25, 0.75), guide='none')+
    scale_size_manual(values=c(2,0.5,1), guide='none')+
    theme_linedraw()+
    theme(legend.position = c(0.85, 0.15) ,legend.direction = 'vertical') +
    labs(color="Features", )+
    xlab("Average log expression") +
    ylab("log ratio of expression") +
    ggtitle(title)
  
  if (add_label){
    plot = plot +
      geom_text_repel(data=FC_df, label = FC_df$label,
                    vjust = vjust, force=force, max.overlaps = 100) }
  
  return(list("plot"=plot, "df"=FC_df)) }


visualize_annotation = function(Annotation_table, celltype){
  Freq_annot = as.data.frame(table(Annotation_table$goslim_goa_description))

  plot = ggplot(Freq_annot, aes(x = reorder(Var1,Freq), y = Freq, fill=Freq)) + 
          geom_col() + 
          coord_flip() +
          ggtitle(paste("Frequence of annotation terms for", celltype ,"geneset"))+
          xlab('GOSLIM GOA annotation terms') + ylab('Frequence')+ labs(fill="Frequence") +
          theme_linedraw() +
          theme(axis.text = element_text(size=12, face = "bold"))
  
  df = rmarkdown::paged_table(unique(Annotation_table[,c("hgnc_symbol","entrezgene_description")])  ) %>% datatable 
  
  return(list("df" = df, "plot"=plot)) }

geneset_to_binary_df = function(geneset_list){
  geneset_list_df = lapply(geneset_list, function(x) as.data.frame(x))
  tmp = ldply(geneset_list_df)
  tmp = reshape2::dcast(tmp, x ~ .id)
  row.names(tmp) = tmp$x
  tmp$x = NULL
  tmp = ifelse(is.na(tmp), 0,1)   
  return( as.data.frame(tmp) ) }

heatmap_from_geneset_list = function(geneset_list, tabula_column, title = " ", 
                                     min_representation = 2, cutree_rows = 1,
                                     clust=TRUE){
  
  n_cut = length(names(geneset_list))
  all_tabula = geneset_list[[tabula_column]]
  all_signature = unique(unlist(geneset_list[ names(geneset_list) != tabula_column ]))
  
  unique_tabula = all_tabula[! all_tabula %in% all_signature]
  Full_geneset = unique(unlist(geneset_list))
  geneset_list[["all"]] = Full_geneset
  df_heatmap = fromList(geneset_list)
  row.names(df_heatmap) = geneset_list$all 
  
  main = paste0(title , "\n", 
        "Presented genes appears in at least ", min_representation, 
        " signature matrix, or in Tabula Sapiens \n",
        length(unique_tabula), 
        " Features uniquely predicted in Tabula sapiens geneset " , 
        "\n 1 : Detected by signature matrix",
        "\n 0 : Not detected by signature matrix")
  
  if (! clust) {return(list("title"=main, "df" = df_heatmap))}
  
  pheatmap::pheatmap(df_heatmap[ 
    rowSums(df_heatmap) > min_representation | df_heatmap[[tabula_column]] == 1 , 
    colnames(df_heatmap) != "all"],
      clustering_distance_rows = "correlation", clustering_distance_cols = "correlation",
      cutree_cols = n_cut, cutree_rows = cutree_rows,
      color = c("#555555","red"), 
      na_col="lightgray",breaks = c(0, .5 , 1),
      fontsize_row = 14, fontsize_col = 15, angle_col = 0, legend_breaks = c(0,1),
      main = main )
}

FC_prop_expr = function(df, n_cell, max_cell){
  
  columns = colnames(df)[str_detect(colnames(df), "prop_expr")]
  A = columns[!str_detect(columns, "vs")]
  B = columns[ str_detect(columns, paste0("other")) ]

  df[[paste0("FC_",A)]] = compute_FC( df[[A]] , df[[B]] , 0)
  return(df) }

kneedle_threshold = function(df, y, sens = 1, min = 0){
  df$value = abs(as.numeric(y))
  df = df[order(df$value, decreasing = TRUE),]
  df$rank = 1:length(y)
  elbow = kneedle::kneedle(x = df$rank, y = df$value, decreasing = TRUE, concave = FALSE, sensitivity = sens)
  
  plot = ggplot(df, aes(x=rank,y=value))+
    geom_line()+
    geom_hline(yintercept = elbow[2], color="blue",alpha=0.5) + geom_vline(xintercept = elbow[1], color="blue",alpha=0.5)+
    ylab("y value") + xlab("Gene Rank")
  
  if (elbow[2] < min) {elbow[2] = min}
  
  return(list("rank" = elbow[1], "threshold" = elbow[2], "plot"=plot )) }

Rprop_plot = function(FC_df, FC_x, y, title = "plot", highlight=c('TOP20'), annotation = "Most Important Features", 
                      vjust = 2, force=3, x_scale = 0, add_label=TRUE, 
                      FC_threshold = 1, prop_threshold=0.25, FC_thresh2 = 1,
                      prop_threshold_to_show = prop_threshold) {
  
  FC_df$selection = ifelse(abs(FC_df[[FC_x]]) > FC_threshold & abs(FC_df[[y]]) > prop_threshold , 
                            yes = "Selected",no = "Not selected")
  Display_2nd_annotation = FALSE
  
  if (FC_threshold <= FC_thresh2){
    FC_df$selection = ifelse(abs(FC_df[[FC_x]]) >= FC_thresh2, yes = "Selected", no = FC_df$selection ) 
    Display_2nd_annotation = TRUE }
  
  selected_number= length(FC_df$selection[FC_df$selection == "Selected"])
    
  if (highlight[1] == "TOP20") {
    
    tmp_df = FC_df[FC_df[[y]] > prop_threshold &  abs(FC_df[[FC_x]]) > FC_threshold , ]
    tmp_df = tmp_df[with(tmp_df, order(abs(tmp_df[[FC_x]]) , decreasing = TRUE)),]
     if ( length(row.names(tmp_df)) > 20) { selected = row.names(tmp_df)[1:20] } 
     else{  selected = row.names(tmp_df) }

    FC_df$selection = ifelse( row.names(FC_df) %in% selected, 
                            yes = "20 Highest FC expressed in most cells",no = FC_df$selection)
    FC_df$label = ifelse(FC_df$selection == "20 Highest FC expressed in most cells", 
                          yes=row.names(FC_df), no="") }
  
  if(highlight[1] == "selection"){
    FC_df$label = ifelse(FC_df$selection == "Selected", 
                          yes=row.names(FC_df), no="") } 
  else{
    FC_df$selection = ifelse(row.names(FC_df) %in% highlight , 
                            yes = annotation, no = FC_df$selection)
    FC_df$label = ifelse(FC_df$selection == annotation, 
                          yes=row.names(FC_df), no="") }
  
  
  maximum_x = max(abs(FC_df[[FC_x]])) + x_scale
  title =paste( c(title, "\n Gene selected :",selected_number),collapse = " " )
  
  plot = ggplot(FC_df, aes(x=FC_df[[FC_x]], y=FC_df[[y]]))+
    geom_point(aes(color=selection, alpha=selection, size = selection))+

    # Annotation of first filter
    annotate("text", y=-0.02, x=-FC_threshold-0.75, 
             label=paste0("log2 FC = -",FC_threshold), 
             colour = "darkblue") +
    geom_vline(xintercept = -FC_threshold, color='darkblue',linetype='dashed') +
    
    annotate("text", y=-0.02, x= FC_threshold+0.75, 
             label=paste0("log2 FC = ",FC_threshold), 
             colour = "darkblue") +
    geom_vline(xintercept = FC_threshold, color='darkblue',linetype='dashed') +
    # Proportion line
    geom_hline(yintercept = prop_threshold, color='darkblue',linetype='dashed') +

    
    coord_cartesian(xlim=c(-maximum_x,maximum_x), ylim=c(-0.025,1)) +
    
    scale_color_manual(values = c("orangered","#777777","#00539CFF"))+
    scale_alpha_manual(values=c(1,0.25, 0.75), guide='none')+
    scale_size_manual(values=c(2,0.5,1), guide='none')+
      
    theme_linedraw()+
    theme(legend.position = c(0.15, 0.85) ,legend.direction = 'vertical') +
    labs(color="Features")+
    ylab("Proportion of cells that express a gene") +
    xlab("log ratio of expression") +
    ggtitle(title)
    
  if (add_label){
    plot = plot +
      geom_text_repel(data=FC_df, label = FC_df$label,
                    vjust = vjust, force=force, max.overlaps = 100) }
      # Annotation of 2nd filter
  if (Display_2nd_annotation){
    plot = plot +
    annotate("text", y=-0.04, x=-FC_thresh2-1, 
             label=paste0("log2 FC = -",round(FC_thresh2, digits = 3)), 
             colour = '#555555') +
    geom_vline(xintercept = -FC_thresh2, color='#555555',linetype='dashed') +
    
    annotate("text", y=-0.04, x= FC_thresh2+1, 
             label=paste0("log2 FC = ",label=round(FC_thresh2, digits = 3)  ), 
             colour = '#555555') +
    geom_vline(xintercept = FC_thresh2, color='#555555',linetype='dashed')  }
  
  if (prop_threshold < 0.1) {
      plot = plot + annotate("text", y=0.05, x= maximum_x*0.75, 
             label=paste0("Proportion of \n cells expressing = ", prop_threshold_to_show), 
             colour = "darkblue") }
  else {
      plot = plot + annotate("text", y=prop_threshold*1.15, x= -maximum_x*0.75, 
             label=paste0("Proportion of cells expressing = ", round(prop_threshold, 3)), 
             colour = "darkblue") }
  
  return(list("plot"=plot, "df"=FC_df)) }


get_permutation_list = function(directory,pattern2,pattern1="Permutation_result_"){
  # Import files
  #directory = file.path(directory)
  tmp = list.files(directory, full.names = TRUE, 
                   recursive = TRUE, pattern=pattern1)

  if (pattern1 == "Permutation_result_") {
    Permutation_list = lapply(tmp, 
                            function(x) x = read.csv2(x, header = TRUE, 
                                            sep = ",", row.names = 1))}
  else {
    Permutation_list = lapply(tmp, function(x) {
      x = read.table(x, header = TRUE , sep=",", row.names=2) } )} 
  
  names(Permutation_list) = tmp
  # Rename each dataframe
  names(Permutation_list) = str_remove_all(
    string = names(Permutation_list) , 
    pattern = paste0(directory , pattern2))
  names(Permutation_list) = str_remove_all(string = names(Permutation_list) , 
                                           pattern =".csv")
  # Change data to numeric
  Permutation_list = lapply(Permutation_list, function(df) {
    df = dplyr::mutate_all(df, function(x) as.numeric(as.character(x) )) } )
  
  # return(list('l'=0,
  #             "dir"=paste0(directory , pattern2),
  #             "names"=names(Permutation_list) ) ) }
  return(Permutation_list)  }


get_genes_from_binary=function(binary_df, expressed_genes=NULL, filter_expressed = FALSE ){
  cell_types = colnames(binary_df)
  cell_geneslist = lapply(cell_types, function(cell){
    #gene_list = list()
    gene_list = row.names(binary_df[ binary_df[[cell]] != 0 , ])
    gene_list = gene_list[gene_list %in% expressed_genes]
    return(gene_list) })
  
  names(cell_geneslist) = cell_types  
  return(cell_geneslist) }


compute_median_expr=function(df,col_to_keep){
  df = df[,col_to_keep]
  df$genes = row.names(df)
  df_melt = reshape2::melt(df)
  return (median(df_melt$value)) }

compute_Avalue=function(expr_A, expr_B){ 
  return(0.5* ( log2(expr_A+1) + log2(expr_B+1) ) ) }
  #return(0.5* ( expr_A + expr_B ) ) }

compute_FC=function(expr_A, expr_B, median_expr){ 
  return(log2( (expr_A+median_expr ) / (expr_B+median_expr) )) }
  #return( (expr_A+median_expr ) / (expr_B+median_expr) ) } 

Get_value_RA = function(df, median_expr){
  A = colnames(df)[!str_detect(colnames(df), "vs|prop_expr")]
  B = colnames(df)[ str_detect(colnames(df), paste0("other_vs_",A)) ]
  B = B[! str_detect( B, "prop_expr")  ]
  df[[paste0("FC_",A)]] = compute_FC( df[[A]] , df[[B]] ,median_expr)
  df[[paste0("Avalue_",A)]] = compute_Avalue( df[[A]] , df[[B]] )
  return(df)
}



build_confusion_heatmap = function(conf_mat,cell_order, 
                                   filter = FALSE, 
                                   column_filter="signature_matrix",
                                   value_filter = "LM22",
                                   direction_color_palette = -1){
  cells_order = as.factor(cell_order)
  cells_order_Y = as.factor(rev(cell_order))
  
  if (direction_color_palette > 0){
    color1 = "white"
    color2 = "black" 
  } else {
    color1 = "black"
    color2 = "white" 
  }
  
  conf_mat$Percentage = round(conf_mat$Percentage*100,2)
  conf_mat$Percentage_show = paste0(as.character(conf_mat$Percentage),"%")
  
  if (filter){ conf_mat = conf_mat[conf_mat[[column_filter]] == value_filter , ] } 
  
  plot = ggplot(data = conf_mat, 
                aes(x=factor(Predicted_label, levels = cells_order), 
                            y=factor(True_label, levels = cells_order_Y), 
                            fill=Percentage)) + 
  geom_tile(color="white")+
  # % cells predicted
  geom_text(data = conf_mat[conf_mat$Percentage > 30,],
    aes(label = Percentage_show), color = color2, size = 3, vjust=-1) +
  geom_text(data = conf_mat[conf_mat$Percentage <= 30,],
    aes(label = Percentage_show), color = color1, size = 3, vjust=-1) +
  
  # Number of cells predicted 
  geom_text(data = conf_mat[conf_mat$Percentage > 30,],
    aes(label = value), color = color2, size = 3, vjust=1) +
  geom_text(data = conf_mat[conf_mat$Percentage <= 30,],
    aes(label = value), color = color1, size = 3, vjust=1) +
  
  xlab("Predicted labels") + ylab("True labels")+
  theme_minimal()+
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))+
  scale_fill_viridis(option = "rocket", direction = direction_color_palette)
  
  return(list("plot"=plot, "conf_mat"=conf_mat)) } 

get_confusion = function(directory,pattern2,pattern1="confusion_heatmap_", 
                         cells ){
  # Import files
  tmp = list.files(directory, full.names = TRUE, 
                   recursive = TRUE, pattern=pattern1)

  conf_heamtap = read.csv2(tmp, header = FALSE, sep = " ", row.names = NULL)
  colnames(conf_heamtap) = cells
  row.names(conf_heamtap) = cells
  conf_mat_percent = conf_heamtap
  
  for (i in seq(1,nrow(conf_heamtap) )) {
    ncell = sum(conf_heamtap[i,])
    conf_mat_percent[i,] = conf_heamtap[i,] / ncell }
  
  colnames(conf_mat_percent) = cells
  row.names(conf_mat_percent) = cells
  conf_mat_percent$True_label = cells
  conf_mat_percent = reshape2::melt(conf_mat_percent)

  colnames(conf_heamtap) = cells
  row.names(conf_heamtap) = cells
  conf_heamtap$True_label = cells
  conf_heamtap = reshape2::melt(conf_heamtap)

  colnames(conf_mat_percent) = c("True_label","Predicted_label","Percentage")
  colnames(conf_heamtap) = c("True_label","Predicted_label","value")
  conf_mat = merge(conf_mat_percent,conf_heamtap)
  
  return(conf_mat)  }
confusion_matrix_setup = function(directory,pattern2,
                                  cells, signature_matrix,
                                  pattern1="confusion_heatmap_"){
  cf = get_confusion(
    directory = directory ,
    pattern2= "/confusion_heatmap_",
    cells = cells)
    
  balanced_acc = round(mean(cf[cf$True_label == cf$Predicted_label,]$Percentage)*100,2)
  cf$subtitle=paste0(signature_matrix, "\n Balanced accuracy = ",balanced_acc,"%")
  cf$signature_matrix = signature_matrix
  return(cf) }

merge_confusion_matrix = function(pattern2, tbs_path,lm22_path,til10_path,cells){
  cf1 = confusion_matrix_setup(
    directory = tbs_path,
    pattern2= "/confusion_heatmap_",
    cells = cells,
    signature_matrix = "Tabula Sapiens")
  cf2 = confusion_matrix_setup(
    directory = lm22_path,
    pattern2= "/confusion_heatmap_",
    cells = cells,
    signature_matrix = "LM22")
  
  cf3 = confusion_matrix_setup(
    directory = til10_path,
    pattern2= "/confusion_heatmap_",
    cells = cells,
    signature_matrix = "TIL10")
  return(rbind(cf1,cf2,cf3)) }


```

# Processing of signature matrix {.tabset}

```{r signature_mat, fig.width=16, fig.height=8, eval=FALSE,warning=FALSE , message=FALSE , message=FALSE}

LM22 = read.csv("/home/acolajanni/Documents/work/doc/LM22/LM22_reduced.txt", sep='\t')
Sigmatrix = read.csv("/home/acolajanni/Documents/work/doc/LM22/Sigmatrix_Binary_reduced.txt", sep='\t')
Vallania = read.csv("/home/acolajanni/Documents/work/doc/LM22/Vallania_Binary_reduced.txt", sep='\t')
TIL10 = read.csv("/home/acolajanni/Documents/work/doc/LM22/TIL10_binary_2sd_reduced.txt", sep='\t')
features_expressed = colnames(read.csv("/home/acolajanni/Documents/work/data/Tabula_sapiens_immune_all/Blood_all_cell_expressed_genes.csv", 
                                       nrows = 1, header = TRUE, sep= '\t'))[2:32285]


LM22_genelist      = get_genes_from_binary(LM22, features_expressed, TRUE)
Sigmatrix_genelist = get_genes_from_binary(Sigmatrix, features_expressed, TRUE)
Vallania_genelist  = get_genes_from_binary(Vallania, features_expressed, TRUE)
TIL10_genelist     = get_genes_from_binary(TIL10, features_expressed, TRUE)


```

# Cell Types

```{r celltypes_og, fig.width=12, fig.height=9,warning=FALSE , message=FALSE}

cellTypes = read.table(file.path(data_dir,"50k_cell_labels.txt"), header=FALSE, sep='\t')
label_count = cellTypes$V1
label_count = as.data.frame(table(label_count) )


ggplot( label_count, aes(x = reorder(label_count,Freq), y = Freq, fill=Freq ) )  +
  geom_col(position = position_dodge(0), width = 0.75) +
    geom_text(color="black",hjust=-0.2, size=3.8, angle=0,
            aes(y = ,label = Freq))+
  coord_flip(ylim = c(0,12000)  ) +
  labs(fill = "Geneset") +
  ylab('Frequency') + xlab('Cell type')+
  ggtitle(paste0("Frequency of cell types in Tabula Sapiens immune (Blood) dataset"))+
  scale_fill_viridis(discrete = FALSE, alpha=1, direction = -1, option='E', name="Absolute frequency") +
  theme_linedraw()+  
  theme( plot.title = element_text(size=15),
         axis.text.y = element_text(size = 12),
         axis.text.x = element_text(size = 12))

```

```{r celltypes, fig.width=15, fig.height=9,warning=FALSE , message=FALSE}

cellTypes = read.table(file.path(data_dir,"Celltypes_of_interest_50k.txt"), header=FALSE, sep='\t')
label_count = cellTypes$V1
label_count = as.data.frame(table(label_count) )
label_count$label_count = ifelse(label_count$label_count == "Plasmocyte",
                                 "Plasma_cell", as.vector(label_count$label_count ) )


ggplot( label_count, aes(x = reorder(label_count,Freq), y = Freq, fill=Freq ) )  +
  geom_col(position = position_dodge(0), width = 0.75) +
    geom_text(color="black",hjust=-0.2, size=3.8, angle=0,
            aes(y = ,label = Freq))+
  coord_flip(ylim = c(0,17000)  ) +
  labs(fill = "Geneset") +
  ylab('Frequency') + xlab('Cell type')+
  ggtitle(paste0("Frequency of cell types in Tabula Sapiens immune (Blood) dataset"))+
  scale_fill_viridis(discrete = FALSE, alpha=1, direction = -1, option='E', name="Absolute frequency") +
  theme_linedraw()+  
  theme( plot.title = element_text(size=15),
         axis.text.y = element_text(size = 12),
         axis.text.x = element_text(size = 12))




n_cells = label_count
row.names(n_cells) = label_count$label_count
n_cells$label_count = NULL
n_cells = as.data.frame(t(n_cells))
n_cells$total = sum(n_cells[1,])
```

```{r celltypes_Liu, fig.width=15, fig.height=5,warning=FALSE , message=FALSE, include=FALSE, eval=TRUE}
data_dir_tmp = "/run/user/1001/gvfs/sftp:host=core.cluster.france-bioinformatique.fr,user=acolajanni/shared/projects/microbiome_translocation/data/scRNAseq/Liu_2021_cell/"

cellTypes = readLines(file.path(data_dir_tmp,"Liu2021_index.txt"))
label_count = as.data.frame(table(cellTypes) )


ggplot( label_count, aes(x = reorder(cellTypes,Freq), y = Freq, fill=Freq ) )  +
  geom_col(position = position_dodge(0), width = 0.75) +
    geom_text(color="black",hjust=-0.2, size=3.8, angle=0,
            aes(y = ,label = Freq))+
  coord_flip(ylim = c(0,40000)  ) +
  labs(fill = "Geneset") +
  ylab('Frequency') + xlab('Cell type')+
  ggtitle(paste0("Frequency of cell types in Liu et al. 2021, Cell. Healthy patients, Innate and Adaptative  "))+
  scale_fill_viridis(discrete = FALSE, alpha=1, direction = -1, option='E', name="Absolute frequency") +
  theme_linedraw()+  
  theme( plot.title = element_text(size=15),
         axis.text.y = element_text(size = 12),
         axis.text.x = element_text(size = 12))

```

```{r celltypes_Liu2, fig.width=15, fig.height=10,warning=FALSE , message=FALSE, include=FALSE}
data_dir_tmp = "/run/user/1001/gvfs/sftp:host=core.cluster.france-bioinformatique.fr,user=acolajanni/shared/projects/microbiome_translocation/data/scRNAseq/Liu_2021_cell/"

cellTypes_innate = read.table(file.path(data_dir_tmp,"Innate_healthy_labels.txt"), header=FALSE, sep='\t')
label_count = cellTypes_innate$V1
label_count = as.data.frame(table(label_count) )
colnames(label_count) = c("cells","Freq")


cellTypes_adaptative = read.table(file.path(data_dir_tmp,"Adaptative_healthy_labels.txt"), header=FALSE, sep='\t')
label_count2 = cellTypes_adaptative$V1
label_count2 = as.data.frame(table(label_count2) )
colnames(label_count2) = c("cells","Freq")

label_count = rbind(label_count,label_count2)

ggplot( label_count, aes(x = reorder(cells,Freq), y = Freq, fill=Freq ) )  +
  geom_col(position = position_dodge(0), width = 0.75) +
    geom_text(color="black",hjust=-0.2, size=3.8, angle=0,
            aes(y = ,label = Freq))+
  coord_flip(ylim = c(0,40000)  ) +
  labs(fill = "Geneset") +
  ylab('Frequency') + xlab('Cell type')+
  ggtitle(paste0("Frequency of cell types in Liu et al. 2021, Cell. Healthy patients, Innate and Adaptative  "))+
  scale_fill_viridis(discrete = FALSE, alpha=1, direction = -1, option='E', name="Absolute frequency") +
  theme_linedraw()+
  theme( plot.title = element_text(size=15),
         axis.text.y = element_text(size = 12),
         axis.text.x = element_text(size = 12))

```

# Filtering genes - Cell vs Cell {.tabset}

```{r LFC, fig.width=8, fig.height=6,warning=FALSE , message=FALSE}
data_dir = file.path(data_dir,"50k_cells_geneset")

groupwise_mean = read.table(file.path( data_dir,"Groupwise_count_non_null_vs_50k.txt"), sep='\t', header = TRUE)
groupwise_prop_full = read.table(file.path( data_dir,"Groupwise_count_non_null_vs_50k.txt"), sep='\t', header = TRUE)


gene_label = readLines(file.path(data_dir,"50k_cell_genes.txt"))
groupwise_mean$genes = gene_label
groupwise_prop_full$genes = gene_label
row.names(groupwise_mean) = gene_label
row.names(groupwise_prop_full) = gene_label

# Compute median expr over all datatable
col = colnames(groupwise_mean)[!str_detect(colnames(groupwise_mean), "vs|genes" )]
expr_median = compute_median_expr(groupwise_mean,col )


colnames(groupwise_prop_full) = paste0(colnames(groupwise_prop_full),"_prop_expr")
groupwise_prop_full$genes = row.names(groupwise_prop_full)
groupwise_mean$genes = row.names(groupwise_mean)
  
groupwise_mean = merge(groupwise_mean, groupwise_prop_full, by='genes')
groupwise_mean$genes_prop_expr = NULL

```




## CD4 vs CD8

**61 Features selected**

```{r LFC_CD84, fig.width=18, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_cd48 = groupwise_mean[str_detect(colnames(groupwise_mean),"CD4|CD8")]
FC_df_cd48$FC_cd48 = log2( (FC_df_cd48$CD4+expr_median ) / (FC_df_cd48$CD8+expr_median) )

elbow=kneedle_threshold(FC_df_cd48, FC_df_cd48$FC_cd48)
elbow$plot
cd48 = Rprop_plot(FC_df_cd48, FC_x = "FC_cd48", y="CD4_prop_expr", 
                title = "Log Fold Change of CD4 vs CD8 depending on the proportion of CD4 cells expressing the genes", 
                #add_label = TRUE, 
                FC_threshold = 1, 
                prop_threshold = 2/n_cells$CD4, 
                prop_threshold_to_show = paste0("2 / ",as.character(n_cells$CD4)),
                highlight = "NCR1",
                FC_thresh2 = elbow$threshold) 

cd84 = Rprop_plot(FC_df_cd48, FC_x = "FC_cd48", y="CD8_prop_expr", 
                title = "Log Fold Change of CD4 vs CD8 depending on the proportion of CD8 cells expressing the genes", 
                #add_label = TRUE, 
                FC_threshold = 1, 
                prop_threshold = 2/n_cells$CD8,
                prop_threshold_to_show = paste0("2 / ",as.character(n_cells$CD8)),
                highlight = "NCR1",
                FC_thresh2 = elbow$threshold) 


cowplot::plot_grid(cd84$plot, cd48$plot, labels = 'AUTO', nrow = 1)



cd84_new = row.names(cd84$df[cd84$df$selection != "Not selected",])
cd48_new = row.names(cd48$df[cd48$df$selection != "Not selected",])


cd48_genes = unique(c(cd84_new,cd48_new))

writeLines(cd48_genes, file.path(data_dir,'CD4vsCD8_3435.txt') )
writeLines(cd48_genes, file.path(git_geneset,'CD4vsCD8_3435.txt') )

```

### 3435 features

```{r cd48_barplot, fig.width=12, fig.height=10,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "CD4-CD8/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_cd48 = Permutation_list$`RF_permutation_First_CD4 vs CD8`
other_cd48$genes = row.names(other_cd48)

cd48_computation = importance_barplot(other_cd48, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 3435 genes selected in CD4 vs CD8 classification")
                                        
cd48_computation$plot

top61_cd48_genes = cd48_computation$ordered_permutation_dataframe[cd48_computation$ordered_permutation_dataframe$mean_importance >0,]$genes 

writeLines(top61_cd48_genes, file.path(data_dir,'CD4vsCD8_61.txt') )
writeLines(top61_cd48_genes, file.path(git_geneset,'CD4vsCD8_61.txt') )

```

## NK vs CD8

**79 Features selected**

```{r LFC_NKCD8, fig.width=18, fig.height=10,  message=F, warning=F}
row.names(groupwise_mean)= groupwise_mean$genes
FC_df_NKcd8 = groupwise_mean[str_detect(colnames(groupwise_mean),"NK|CD8")]
FC_df_NKcd8$FC_NKCD8 = log2( (FC_df_NKcd8$NK_cell+expr_median ) / (FC_df_NKcd8$CD8+expr_median) )
FC_df_NKcd8$Avalue = 0.5* ( log2(FC_df_NKcd8$NK_cell+1) + log2(FC_df_NKcd8$CD8+1) )

elbow=kneedle_threshold(FC_df_NKcd8, FC_df_NKcd8$FC_NKCD8)
elbow$plot

nkcd8 = Rprop_plot(FC_df_NKcd8, FC_x = "FC_NKCD8", y="NK_cell_prop_expr", 
                title = "Log Fold Change of NK vs CD8 depending on the proportion of NK cells expressing the genes", 
                #add_label = TRUE, 
                FC_threshold = 1, 
                prop_threshold = 2/n_cells$NK_cell,
                prop_threshold_to_show = paste0("2 / ",as.character(n_cells$NK_cell)),
                #highlight = test,
                FC_thresh2 = elbow$threshold) 

cd8nk = Rprop_plot(FC_df_NKcd8, FC_x = "FC_NKCD8", y="CD8_prop_expr", 
                title = "Log Fold Change of NK vs CD8 depending on the proportion of CD8 cells expressing the genes", 
                #add_label = TRUE, 
                FC_threshold = 1, 
                prop_threshold = 2/n_cells$CD8,
                prop_threshold_to_show = paste0("2 / ",as.character(n_cells$CD8)),
                #highlight = test,
                FC_thresh2 = elbow$threshold) 



cowplot::plot_grid(nkcd8$plot, cd8nk$plot, labels = 'AUTO', nrow = 1)

nkcd8_new = row.names(nkcd8$df[nkcd8$df$selection != "Not selected",])
cd8nk_new = row.names(cd8nk$df[cd8nk$df$selection != "Not selected",])

cd8nk_genes = unique(c(nkcd8_new,cd8nk_new))

writeLines(cd8nk_genes, file.path(data_dir,'CD8vsNK_1028.txt') )
writeLines(cd8nk_genes, file.path(git_geneset,'CD8vsNK_1028.txt') )

```

### 1028 features

```{r nkcd8_barplot, fig.width=12, fig.height=10,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "NK-CD8/" ),
  pattern2= "/RF_permutation/Permutation_result_")


nkcd8 = Permutation_list$`RF_permutation_First_NK vs CD8`
nkcd8$genes = row.names(nkcd8)


nkcd8_computation = importance_barplot(nkcd8, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 1028 genes selected in NK vs CD8 classification")
                                        
nkcd8_computation$plot+
   theme(axis.text.y=element_blank(),
        axis.ticks.y=element_blank()
        )

top247_nkcd8_genes = nkcd8_computation$ordered_permutation_dataframe[nkcd8_computation$ordered_permutation_dataframe$mean_importance >0,]$genes 
writeLines(top247_nkcd8_genes, file.path(data_dir,'NKvsCD8_247.txt') )
writeLines(top247_nkcd8_genes, file.path(git_geneset,'NKvsCD8_247.txt') )

```

### 247 features

```{r nkcd8_barplot248, fig.width=12, fig.height=10,warning=FALSE , message=FALSE}

nkcd8_2 = Permutation_list$`RF_permutation_Second_NK vs CD8`
nkcd8_2$genes = row.names(nkcd8_2)

nkcd8_computation = importance_barplot(nkcd8_2, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 247 genes selected in NK vs CD8 classification")
                                        
nkcd8_computation$plot

top79_nkcd8_genes = nkcd8_computation$ordered_permutation_dataframe[nkcd8_computation$ordered_permutation_dataframe$mean_importance >0,]$genes 
writeLines(top79_nkcd8_genes, file.path(data_dir,'NKvsCD8_79.txt') )
writeLines(top79_nkcd8_genes, file.path(git_geneset,'NKvsCD8_79.txt') )

```

## Naive vs Memory Bcells

**42 Features selected**

```{r LFC_naivememory_bcells, fig.width=18, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_naiveMem = groupwise_mean[str_detect(colnames(groupwise_mean),"naive|memory")]

FC_df_naiveMem$FC_nmB = log2( 
  (FC_df_naiveMem$memory_B_cell+expr_median ) / (FC_df_naiveMem$naive_B_cell+expr_median) )

elbow=kneedle_threshold(FC_df_naiveMem, FC_df_naiveMem$FC_nmB)
elbow$plot

naivemem = Rprop_plot(FC_df_naiveMem, FC_x = "FC_nmB", y="naive_B_cell_prop_expr", 
                title = "Log Fold Change of Naive vs Memory Bcells depending on the proportion of naive Bcells expressing the genes", 
                #add_label = TRUE, 
                FC_threshold = 1, 
                prop_threshold = 2/n_cells$naive_B_cell,
                prop_threshold_to_show = paste0("2 / ",as.character(n_cells$naive_B_cell)),
                FC_thresh2 = elbow$threshold) 

memnaive = Rprop_plot(FC_df_naiveMem, FC_x = "FC_nmB", y="memory_B_cell_prop_expr", 
                title = "Log Fold Change of Naive vs Memory Bcells depending on the proportion of memory Bcells expressing the genes", 
                #add_label = TRUE, 
                FC_threshold = 1, 
                prop_threshold = 2/n_cells$memory_B_cell, 
                prop_threshold_to_show = paste0("2 / ",as.character(n_cells$memory_B_cell)),
                FC_thresh2 = elbow$threshold) 


cowplot::plot_grid(naivemem$plot, memnaive$plot, labels = 'AUTO', nrow = 1)


memnaive_new = row.names(memnaive$df[memnaive$df$selection != "Not selected",])
naivemem_new = row.names(naivemem$df[naivemem$df$selection != "Not selected",])
nmBcell_genes = unique(c(memnaive_new,naivemem_new))


writeLines(nmBcell_genes, file.path(data_dir,'naiveVSmemory_B_3582.txt') )
writeLines(nmBcell_genes, file.path(git_geneset,'naiveVSmemory_B_3582.txt') )

```

### 3582 features

```{r B3_barplot, fig.width=12, fig.height=10,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Memory-Naive_Bcell/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_nm = Permutation_list$`RF_permutation_First_Naive vs Memory`
other_nm$genes = row.names(other_nm)

nm_computation = importance_barplot(other_nm, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 3582 genes selected in Memory vs Naive Bcells classification")
                                        
nm_computation$plot

top42_naive_mem = nm_computation$ordered_permutation_dataframe[nm_computation$ordered_permutation_dataframe$mean_importance >0,]$genes 

writeLines(top42_naive_mem, file.path(data_dir,'naiveVSmemory_B_42.txt') )
writeLines(top42_naive_mem, file.path(git_geneset,'naiveVSmemory_B_42.txt') )

```

# Filtering genes - Cell vs all {.tabset}

## NK vs other

**178 Features selected**

```{r LFC_NK, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_NK = groupwise_mean[str_detect(colnames(groupwise_mean),"NK")]
FC_df_NK = Get_value_RA(FC_df_NK, expr_median)
FC_df_NK = FC_prop_expr(FC_df_NK)


nk = RA_plot(FC_df_NK, "Avalue_NK_cell", "FC_NK_cell", title = "RA plot NK vs other", add_label = FALSE)



elbow=kneedle_threshold(FC_df_NK, FC_df_NK$FC_NK_cell)
elbow$plot

nk = Rprop_plot(FC_df_NK,FC_x = "FC_NK_cell", y="NK_cell_prop_expr", 
                title = "Log Fold Change of NK vs other depending on the proportion of NK cells expressing the genes", 
                highlight = "KLRB1",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 


nk$plot


nk_new = row.names(nk$df[nk$df$selection != "Not selected",])


writeLines(nk_new, file.path(data_dir,'NK_926.txt') )
writeLines(nk_new, file.path(git_geneset,'NK_926.txt') )

```

### 926 features

```{r nk_barplot, fig.width=12, fig.height=10,warning=FALSE , message=FALSE}
Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "NK/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_NK = Permutation_list$`RF_permutation_First_NK vs other`
other_NK$genes = row.names(other_NK)

NK_computation = importance_barplot(other_NK, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 926 genes selected in Other vs NK cell classification", 
                                    conf_interval = TRUE) 
                                        
NK_computation$plot  +
   theme(axis.text.y=element_blank(),
        axis.ticks.y=element_blank()
        )

top178_nk = NK_computation$ordered_permutation_dataframe[NK_computation$ordered_permutation_dataframe$mean_importance >0,]$genes 
writeLines(top178_nk, file.path(data_dir,'NK_178.txt') )
writeLines(top178_nk, file.path(git_geneset,'NK_178.txt') )

```

### Performances

```{r heatmap_nk, fig.width=14,fig.height=11,message=FALSE}
path_heatmap = file.path(path,"results/Liu2021/Signature_matrix_v2/")

cf_Liu2021 = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k/NK_cell/" ),
  til10_path = file.path(path_heatmap, "TIL10/NK_cell/RF/" ),
  lm22_path = file.path(path_heatmap, "LM22/NK_cell/RF/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","NK_cell"))

path_heatmap = file.path(path,"results/Tabula_sapiens/Signature_matrix_v2/")

cf_tbs = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/NK_cell/" ),
  til10_path = file.path(path_heatmap, "TIL10/NK_cell/" ),
  lm22_path = file.path(path_heatmap, "LM22/NK_cell/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","NK_cell"))

cf_tbs$dataset = "Tabula Sapiens data"
cf_Liu2021$dataset = "Liu 2021 Cell data"
cf = rbind(cf_tbs,cf_Liu2021)

comp = build_confusion_heatmap(cf,  
                               c("Other_cells","NK_cell"),
                               filter = FALSE,
                               direction_color_palette = 1)
comp$plot + theme_dark() + 
  facet_wrap(. ~ dataset+subtitle, ncol=3) + 
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),

        
        plot.title = element_text(hjust = 0.5, size = 16,face="bold"))+
        
  ggtitle("Signature matrix prediction performance with Random Forest - NK cell Classification") 
```

## CD8 vs other

**79 Features selected**

```{r LFC_CD8, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_CD8 = groupwise_mean[str_detect(colnames(groupwise_mean),"CD8")]
FC_df_CD8 = Get_value_RA(FC_df_CD8, expr_median)
FC_df_CD8 = FC_prop_expr(FC_df_CD8)


elbow=kneedle_threshold(FC_df_CD8, FC_df_CD8$FC_CD8) 
elbow$plot

cd8 = Rprop_plot(FC_df_CD8,FC_x = "FC_CD8", y="CD8_prop_expr", 
                title = "Log Fold Change of CD8 vs other depending on the proportion of CD8 cells expressing the genes",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

cd8$plot


cd8_new = row.names(cd8$df[cd8$df$selection != "Not selected",])


writeLines(cd8_new, file.path(data_dir,'CD8_639.txt') )
writeLines(cd8_new, file.path(git_geneset,'CD8_639.txt') )

```

### 639 features

```{r cd8_barplot309, fig.width=12, fig.height=15,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "CD8/" ),
  pattern2= "/RF_permutation/Permutation_result_")


other_CD8 = Permutation_list$`RF_permutation_First_CD8 vs other`
other_CD8$genes = row.names(other_CD8)

cd8_computation = importance_barplot(other_CD8, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 639 genes selected in Other vs CD8 classification")
                                        
cd8_computation$plot+ 
  theme(axis.text.y = element_text(size=10))


top79_cd8 = cd8_computation$ordered_permutation_dataframe[cd8_computation$ordered_permutation_dataframe$mean_importance >0,]$genes 
writeLines(top79_cd8, file.path(data_dir,'CD8_79.txt') )
writeLines(top79_cd8, file.path(git_geneset,'CD8_79.txt') )

```

### Performances

```{r heatmap_cd8, fig.width=14,fig.height=11,message=FALSE}
path_heatmap = file.path(path,"results/Liu2021/Signature_matrix_v2/")

cf_Liu2021 = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/CD8/" ),
  til10_path = file.path(path_heatmap, "TIL10/CD8/RF/" ),
  lm22_path = file.path(path_heatmap, "LM22/CD8/RF/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","CD8"))

path_heatmap = file.path(path,"results/Tabula_sapiens/Signature_matrix_v2/")

cf_tbs = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/CD8/" ),
  til10_path = file.path(path_heatmap, "TIL10/CD8/" ),
  lm22_path = file.path(path_heatmap, "LM22/CD8/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","CD8"))

cf_tbs$dataset = "Tabula Sapiens data"
cf_Liu2021$dataset = "Liu 2021 Cell data"
cf = rbind(cf_tbs,cf_Liu2021)

comp = build_confusion_heatmap(cf,  
                               c("Other_cells","CD8"),
                               filter = FALSE,
                               direction_color_palette = 1)
comp$plot + theme_dark() + 
  facet_wrap(. ~ dataset+subtitle, ncol=3) + 
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),

        
        plot.title = element_text(hjust = 0.5, size = 16,face="bold"))+
        
  ggtitle("Signature matrix prediction performance with Random Forest - CD8 Classification") 
```

## CD4 vs other

**114 Features selected**

```{r LFC_CD4, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_CD4 = groupwise_mean[str_detect(colnames(groupwise_mean),"CD4")]
FC_df_CD4 = Get_value_RA(FC_df_CD4, expr_median)
FC_df_CD4 = FC_prop_expr(FC_df_CD4)


elbow=kneedle_threshold(FC_df_CD4, FC_df_CD4$FC_CD4) 
elbow$plot

cd4 = Rprop_plot(FC_df_CD4,FC_x = "FC_CD4", y="CD4_prop_expr", 
                title = "Log Fold Change of CD4 vs other depending on the proportion of CD4 cells expressing the genes",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

cd4$plot


cd4_new = row.names(cd4$df[cd4$df$selection != "Not selected",])


writeLines(cd4_new, file.path(data_dir,'CD4_1402.txt') )
writeLines(cd4_new, file.path(git_geneset,'CD4_1402.txt') )

```

### 1402 features

```{r cd4_barplot, fig.width=12, fig.height=14,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "CD4/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_CD4 = Permutation_list$`RF_permutation_First_CD4 vs other`
other_CD4$genes = row.names(other_CD4)

cd4_computation = importance_barplot(other_CD4, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 1402 genes selected in Other vs CD4 classification")
                                        
cd4_computation$plot+ 
  theme(axis.text.y = element_text(size=0))


CD4_top319 = cd4_computation$ordered_permutation_dataframe[cd4_computation$ordered_permutation_dataframe$mean_importance > 0,]$genes

writeLines(CD4_top319, file.path(data_dir,'CD4_319.txt') )
writeLines(CD4_top319, file.path(git_geneset,'CD4_319.txt') )

```

### 319 features

```{r cd4_barplot497, fig.width=18, fig.height=15,warning=FALSE , message=FALSE}

other_CD4 = Permutation_list$`RF_permutation_Second_CD4 vs other`
other_CD4$genes = row.names(other_CD4)

cd4_computation = importance_barplot(other_CD4, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 319 genes selected in Other vs CD4 classification")
                                        
cd4_computation$plot+ 
  theme(axis.text.y = element_text(size=8))


CD4_top114 = cd4_computation$ordered_permutation_dataframe[cd4_computation$ordered_permutation_dataframe$mean_importance > 0,]$genes
writeLines(CD4_top114, file.path(data_dir,'CD4_114.txt') )
writeLines(CD4_top114, file.path(git_geneset,'CD4_114.txt') )

```

### Performances

```{r heatmap_cd4, fig.width=14,fig.height=11,message=FALSE}
path_heatmap = file.path(path,"results/Liu2021/Signature_matrix_v2/")

cf_Liu2021 = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/CD4/" ),
  til10_path = file.path(path_heatmap, "TIL10/CD4/RF/" ),
  lm22_path = file.path(path_heatmap, "LM22/CD4/RF/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","CD4"))

path_heatmap = file.path(path,"results/Tabula_sapiens/Signature_matrix_v2/")

cf_tbs = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/CD4/" ),
  til10_path = file.path(path_heatmap, "TIL10/CD4/" ),
  lm22_path = file.path(path_heatmap, "LM22/CD4/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","CD4"))

cf_tbs$dataset = "Tabula Sapiens data"
cf_Liu2021$dataset = "Liu 2021 Cell data"
cf = rbind(cf_tbs,cf_Liu2021)

comp = build_confusion_heatmap(cf,  
                               c("Other_cells","CD4"),
                               filter = FALSE,
                               direction_color_palette = 1)
comp$plot + theme_dark() + 
  facet_wrap(. ~ dataset+subtitle, ncol=3) + 
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),

        
        plot.title = element_text(hjust = 0.5, size = 16,face="bold"))+
        
  ggtitle("Signature matrix prediction performance with Random Forest - CD4 Classification") 
```

## Naive Bcell vs other

**29 Features selected**

```{r LFC_naive_bcells, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_naiveB = groupwise_mean[str_detect(colnames(groupwise_mean),"naive")]
FC_df_naiveB = Get_value_RA(FC_df_naiveB, expr_median)
FC_df_naiveB = FC_prop_expr(FC_df_naiveB)


elbow=kneedle_threshold(FC_df_naiveB, FC_df_naiveB$FC_naive_B_cell, min = 1) 

elbow$plot

naiveB = Rprop_plot(FC_df_naiveB,FC_x = "FC_naive_B_cell", y="naive_B_cell_prop_expr", 
                title = "Log Fold Change of Naive B cell vs other depending on the proportion of CD8 cells expressing the genes",
                add_label = FALSE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

naiveB$plot


naiveB_genes = row.names(naiveB$df[naiveB$df$selection != "Not selected",])

writeLines(naiveB_genes, file.path(data_dir,'naiveB_1115.txt') )
writeLines(naiveB_genes, file.path(git_geneset,'naiveB_1115.txt') )

```

### 1115 features

```{r naiveB_barplot, fig.width=12, fig.height=8,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Naive_Bcell/" ),
  pattern2= "/RF_permutation/Permutation_result_RF_permutation_erratum_")


other_naiveB = Permutation_list$`naive_B_cell vs other`
other_naiveB$genes = row.names(other_naiveB)

naiveB_computation = importance_barplot(other_naiveB, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 1115 genes selected in Other vs naive Bcell classification",
                                        conf_interval = FALSE) 
  
                                        
naiveB_computation$plot + 
  theme(axis.text.y=element_blank(),
        axis.ticks.y=element_blank() )

naiveB_top347 = naiveB_computation$ordered_permutation_dataframe[naiveB_computation$ordered_permutation_dataframe$mean_importance > 0,]$genes
writeLines(naiveB_top347, file.path(data_dir,'naiveB_347.txt') )
writeLines(naiveB_top347, file.path(git_geneset,'naiveB_347.txt') )

```

### 347 features

```{r naiveB_barplot2, fig.width=12, fig.height=8,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Naive_Bcell/" ),
  pattern2= "/RF_permutation_second/Permutation_result_RF_permutation_erratum_")


other_naiveB = Permutation_list$`second_naive_B_cell vs other`
other_naiveB$genes = row.names(other_naiveB)

naiveB_computation = importance_barplot(other_naiveB, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 347 genes selected in Other vs naive Bcell classification",
                                        conf_interval = FALSE) 
  
                                        
naiveB_computation$plot + 
  theme(axis.text.y = element_text(size=10))


naiveB_top17 = naiveB_computation$ordered_permutation_dataframe[naiveB_computation$ordered_permutation_dataframe$mean_importance > 0,]$genes
writeLines(naiveB_top17, file.path(data_dir,'naiveB_17.txt') )
writeLines(naiveB_top17, file.path(git_geneset,'naiveB_17.txt') )

# naiveB_29 = readLines("/home/acolajanni/Documents/work/GITHUB/save_erratum_22-08/naiveB_29.txt")
# upset(fromList(list("29" = naiveB_29, "17" = naiveB_top17)))

```


### Performances

```{r heatmap_naive, fig.width=14,fig.height=11,message=FALSE}
path_heatmap = file.path(path,"results/Liu2021/Signature_matrix_v2/")

cf_Liu2021 = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/naive_B_cell/" ),
  til10_path = file.path(path_heatmap, "TIL10/naive_B_cell/RF/" ),
  lm22_path = file.path(path_heatmap, "LM22/naive_B_cell/RF/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","naive_B_cell"))

path_heatmap = file.path(path,"results/Tabula_sapiens/Signature_matrix_v2/")

cf_tbs = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/naive_B_cell/" ),
  til10_path = file.path(path_heatmap, "TIL10/naive_B_cell/" ),
  lm22_path = file.path(path_heatmap, "LM22/naive_B_cell/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","naive_B_cell"))

cf_tbs$dataset = "Tabula Sapiens data"
cf_Liu2021$dataset = "Liu 2021 Cell data"
cf = rbind(cf_tbs,cf_Liu2021)

comp = build_confusion_heatmap(cf,  
                               c("Other_cells","naive_B_cell"),
                               filter = FALSE,
                               direction_color_palette = 1)
comp$plot + theme_dark() + 
  facet_wrap(. ~ dataset+subtitle, ncol=3) + 
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),

        
        plot.title = element_text(hjust = 0.5, size = 16,face="bold"))+
        
  ggtitle("Signature matrix prediction performance with Random Forest - naive B cell Classification") 
```

## Memory Bcell vs other

**53 Features selected**

```{r LFC_mem_bcells, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_memB = groupwise_mean[str_detect(colnames(groupwise_mean),"memory")]
FC_df_memB = Get_value_RA(FC_df_memB, expr_median)
FC_df_memB = FC_prop_expr(FC_df_memB)


elbow=kneedle_threshold(FC_df_memB, FC_df_memB$FC_memory_B_cell,min = 1) 
elbow$plot
memB = Rprop_plot(FC_df_memB,FC_x = "FC_memory_B_cell", y="memory_B_cell_prop_expr", 
                title = "Log Fold Change of Memory B cell vs other depending on the proportion of CD8 cells expressing the genes",
                highlight = "NCR1",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

memB$plot

memB_genes = row.names(memB$df[memB$df$selection != "Not selected",])

writeLines(memB_genes, file.path(data_dir,'memoryB_1056.txt') )
writeLines(memB_genes, file.path(git_geneset,'memoryB_1056.txt') )

```

### 1056 features

```{r memB_barplot, fig.width=12, fig.height=10,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Memory_Bcell/" ),
  pattern2= "/RF_permutation/Permutation_result_RF_permutation_erratum_")

# other_memB = lapply(Permutation_list, function(x) {return(x[,1:(ncol(x)-2)]) } ) 
# other_memB = rlist::list.cbind(other_memB)
# mean_imp = rowMeans(other_memB)
# other_memB$genes = row.names(other_memB) 
# row_stdev <- apply(other_memB, 1, sd, na.rm=TRUE)
# 
# other_memB$mean_importance = mean_imp
# other_memB$std_importance = row_stdev

other_memB = Permutation_list$`memory_B_cell vs other`
other_memB$Genes = row.names(other_memB) 
  
memB_computation = importance_barplot(other_memB, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 1056 genes selected in Other vs nnaive Bcell classification")
                                        
memB_computation$plot + 
  theme(axis.text.y=element_blank(),
        axis.ticks.y=element_blank() )

top588_memB = memB_computation$ordered_permutation_dataframe[memB_computation$ordered_permutation_dataframe$mean_importance > 0 ,]$genes

writeLines(top588_memB, file.path(data_dir,'memoryB_588.txt') )
writeLines(top588_memB, file.path(git_geneset,'memoryB_588.txt') )

```

### 588 features

```{r memB_barplot2, fig.width=12, fig.height=8,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Memory_Bcell/" ),
  pattern2= "/RF_permutation_second/Permutation_result_RF_permutation_erratum_")


other_memB = Permutation_list$`second_memory_B_cell vs other`
other_memB$genes = row.names(other_memB)

memB_computation = importance_barplot(other_memB, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 588 genes selected in Other vs memory Bcell classification",
                                        conf_interval = FALSE) 
  
                                        
memB_computation$plot + 
  theme(axis.text.y = element_text(size=5))


top147_memB = memB_computation$ordered_permutation_dataframe[memB_computation$ordered_permutation_dataframe$mean_importance > 0 ,]$genes

writeLines(top147_memB, file.path(data_dir,'memoryB_147.txt') )
writeLines(top147_memB, file.path(git_geneset,'memoryB_147.txt') )

```

### Performances

```{r heatmap_memo, fig.width=14,fig.height=11,message=FALSE}
path_heatmap = file.path(path,"results/Liu2021/Signature_matrix_v2/")

cf_Liu2021 = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/memory_B_cell/" ),
  til10_path = file.path(path_heatmap, "TIL10/memory_B_cell/RF/" ),
  lm22_path = file.path(path_heatmap, "LM22/memory_B_cell/RF/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","memory_B_cell"))

path_heatmap = file.path(path,"results/Tabula_sapiens/Signature_matrix_v2/")

cf_tbs = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/memory_B_cell/" ),
  til10_path = file.path(path_heatmap, "TIL10/memory_B_cell/" ),
  lm22_path = file.path(path_heatmap, "LM22/memory_B_cell/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","memory_B_cell"))

cf_tbs$dataset = "Tabula Sapiens data"
cf_Liu2021$dataset = "Liu 2021 Cell data"
cf = rbind(cf_tbs,cf_Liu2021)

comp = build_confusion_heatmap(cf,  
                               c("Other_cells","memory_B_cell"),
                               filter = FALSE,
                               direction_color_palette = 1)
comp$plot + theme_dark() + 
  facet_wrap(. ~ dataset+subtitle, ncol=3) + 
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),

        
        plot.title = element_text(hjust = 0.5, size = 16,face="bold"))+
        
  ggtitle("Signature matrix prediction performance with Random Forest - memory B cell Classification") 
```

## Neutrophils vs other

**84 Features selected**

```{r LFC_neutro, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_neutro = groupwise_mean[str_detect(colnames(groupwise_mean),"Neutro")]
FC_df_neutro = Get_value_RA(FC_df_neutro, expr_median)
FC_df_neutro = FC_prop_expr(FC_df_neutro)


elbow=kneedle_threshold(FC_df_neutro, FC_df_neutro$FC_Neutrophil) 
elbow$plot

neutro = Rprop_plot(FC_df_neutro,FC_x = "FC_Neutrophil", y="Neutrophil_prop_expr", 
                title = "Log Fold Change of Neutrophil vs other depending on the proportion of Neutrophil cells expressing the genes",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

neutro$plot


neutro_new = row.names(neutro$df[neutro$df$selection != "Not selected",])


writeLines(neutro_new, file.path(data_dir,'Neutro_825.txt') )
writeLines(neutro_new, file.path(git_geneset,'Neutro_825.txt') )

```

### 825 features

```{r neutro_barplot, fig.width=12, fig.height=15,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Neutrophil/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_neutro = Permutation_list$`RF_permutation_First_Neutrophil vs other`
other_neutro$genes = row.names(other_neutro)

neutro_computation = importance_barplot(other_neutro, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 825 genes selected in Other vs neutrophil classification")
                                        
neutro_computation$plot + 
  theme(axis.text.y = element_text(size=10))

neutro_top84 = neutro_computation$ordered_permutation_dataframe[neutro_computation$ordered_permutation_dataframe$mean_importance>0,]$genes
writeLines(neutro_top84, file.path(data_dir,'Neutro_84.txt') )
writeLines(neutro_top84, file.path(git_geneset,'Neutro_84.txt') )

```

### Performances - No neutrophils in Liu2021 data

```{r heatmap_neutro, fig.width=14,fig.height=6,message=FALSE}


path_heatmap = file.path(path,"results/Tabula_sapiens/Signature_matrix_v2/")

cf_tbs = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/Neutrophil/" ),
  til10_path = file.path(path_heatmap, "TIL10/Neutrophil/" ),
  lm22_path = file.path(path_heatmap, "LM22/Neutrophil/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","Neutrophil"))

cf_tbs$dataset = "Tabula Sapiens data"

comp = build_confusion_heatmap(cf_tbs,  
                               c("Other_cells","Neutrophil"),
                               filter = FALSE,
                               direction_color_palette = 1)
comp$plot + theme_dark() + 
  facet_wrap(. ~ dataset+subtitle, ncol=3) + 
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),

        
        plot.title = element_text(hjust = 0.5, size = 16,face="bold"))+
        
  ggtitle("Signature matrix prediction performance with Random Forest - Neutrophils Classification") 
```

## Plasmocyte vs other

**19 Features selected**

```{r LFC_plasmo, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_Plasmo = groupwise_mean[str_detect(colnames(groupwise_mean),"Plasmo")]
FC_df_Plasmo = Get_value_RA(FC_df_Plasmo, expr_median)
FC_df_Plasmo = FC_prop_expr(FC_df_Plasmo)


elbow=kneedle_threshold(FC_df_Plasmo, FC_df_Plasmo$FC_Plasmocyte) 
elbow$plot

plasmo = Rprop_plot(FC_df_Plasmo,FC_x = "FC_Plasmocyte", y="Plasmocyte_prop_expr", 
                title = "Log Fold Change of Plasmocyte vs other depending on the proportion of Plasmocyte cells expressing the genes",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

plasmo$plot


plasmo_new = row.names(plasmo$df[plasmo$df$selection != "Not selected",])


writeLines(plasmo_new, file.path(data_dir,'Plasmocyte_5172.txt') )
writeLines(plasmo_new, file.path(git_geneset,'Plasmocyte_5172.txt') )

```

### 5172 features

```{r plasmo_barplot, fig.width=12, fig.height=7,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Plasmocyte/" ),
  pattern2= "/RF_permutation/Permutation_result_")


other_plasmo = Permutation_list$`RF_permutation_First_Plasmocyte vs other`

plasmo_computation = importance_barplot(other_plasmo, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 5172 genes selected in Other vs Plasmocyte classification")
                                        
plasmo_computation$plot + 
  theme(axis.text.y = element_text(size=10))

plasmo_top19 = plasmo_computation$ordered_permutation_dataframe[plasmo_computation$ordered_permutation_dataframe$mean_importance>0,]$genes
writeLines(plasmo_top19, file.path(data_dir,'Plasmocyte_19.txt') )
writeLines(plasmo_top19, file.path(git_geneset,'Plasmocyte_19.txt') )

```

### Performances - No Plasmocytes in Liu2021 data / TIL10

```{r heatmap_plasmo, fig.width=14,fig.height=9,message=FALSE}

path_heatmap = file.path(path,"results/Tabula_sapiens/Signature_matrix_v2/")

cf1 = confusion_matrix_setup(
    directory = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/Plasmocyte/" ),
    pattern2= "/confusion_heatmap_",
    cells = c("Other_cells","Plasmocyte"),
    signature_matrix = "Tabula Sapiens")

cf2 = confusion_matrix_setup(
    directory = file.path(path_heatmap, "LM22/Plasmocyte/" ),
    pattern2= "/confusion_heatmap_",
    cells = c("Other_cells","Plasmocyte"),
    signature_matrix = "LM22")

cf_tbs = rbind(cf1,cf2)
cf_tbs$dataset = "Tabula Sapiens data"

comp = build_confusion_heatmap(cf_tbs,  
                               c("Other_cells","Plasmocyte"),
                               filter = FALSE,
                               direction_color_palette = 1)
comp$plot + theme_dark() + 
  facet_wrap(. ~ dataset+subtitle, ncol=3) + 
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),

        
        plot.title = element_text(hjust = 0.5, size = 16,face="bold"))+
        
  ggtitle("Signature matrix prediction performance with Random Forest - Plasmocytes Classification") 
```

## Monocyte vs other

**174 Features selected**

```{r LFC_mono, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_mono = groupwise_mean[str_detect(colnames(groupwise_mean),"Mono")]
FC_df_mono = Get_value_RA(FC_df_mono, expr_median)
FC_df_mono = FC_prop_expr(FC_df_mono)


elbow=kneedle_threshold(FC_df_mono, FC_df_mono$FC_Monocyte) 
elbow$plot

mono = Rprop_plot(FC_df_mono,FC_x = "FC_Monocyte", y="Monocyte_prop_expr", 
                title = "Log Fold Change of Monocyte vs other depending on the proportion of Monocyte cells expressing the genes",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

mono$plot


mono_new = row.names(mono$df[mono$df$selection != "Not selected",])


writeLines(mono_new, file.path(data_dir,'Monocyte_4022.txt') )
writeLines(mono_new, file.path(git_geneset,'Monocyte_4022.txt') )

```

### 4022 features

```{r mono_barplot4022, fig.width=12, fig.height=14,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Monocyte/" ),
  pattern2= "/RF_permutation/Permutation_result_")


other_mono = Permutation_list$`RF_permutation_First_Monocyte vs other`

mono_computation = importance_barplot(other_mono, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 4022 genes selected in Other vs Monocyte classification")
                                        
mono_computation$plot + 
  theme(axis.text.y = element_text(size=0))


mono_top1870 = mono_computation$ordered_permutation_dataframe[mono_computation$ordered_permutation_dataframe$mean_importance > 0,]$genes
writeLines(mono_top1870, file.path(git_geneset,'Monocyte_1870.txt') )
writeLines(mono_top1870, file.path(data_dir,'Monocyte_1870.txt') )

```

### 1870 features

```{r mono_barplot1870, fig.width=12, fig.height=14,warning=FALSE , message=FALSE}

other_mono = Permutation_list$`RF_permutation_second_Monocyte vs other`



mono_computation = importance_barplot(other_mono, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 4022 genes selected in Other vs Monocyte classification")
                                        
mono_computation$plot + 
  theme(axis.text.y = element_text(size=0))


mono_top369 = mono_computation$ordered_permutation_dataframe[mono_computation$ordered_permutation_dataframe$mean_importance > 0,]$genes
writeLines(mono_top369, file.path(git_geneset,'Monocyte_369.txt') )
writeLines(mono_top369, file.path(data_dir,'Monocyte_369.txt') )

```

### 369 features

```{r mono_barplot369, fig.width=12, fig.height=14,warning=FALSE , message=FALSE}

other_mono = Permutation_list$`RF_permutation_third_Monocyte vs other`


mono_computation = importance_barplot(other_mono, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 4022 genes selected in Other vs Monocyte classification")
                                        
mono_computation$plot + 
  theme(axis.text.y = element_text(size=5))


mono_top174 = mono_computation$ordered_permutation_dataframe[mono_computation$ordered_permutation_dataframe$mean_importance > 0,]$genes
writeLines(mono_top174, file.path(git_geneset,'Monocyte_174.txt') )
writeLines(mono_top174, file.path(data_dir,'Monocyte_174.txt') )

```

### Performances

```{r heatmap_mono, fig.width=14,fig.height=11,message=FALSE}
path_heatmap = file.path(path,"results/Liu2021/Signature_matrix_v2/")

cf_Liu2021 = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/Monocyte/" ),
  til10_path = file.path(path_heatmap, "TIL10/Monocyte/RF/" ),
  lm22_path = file.path(path_heatmap, "LM22/Monocyte/RF/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","Monocyte"))

path_heatmap = file.path(path,"results/Tabula_sapiens/Signature_matrix_v2/")

cf_tbs = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/Monocyte/" ),
  til10_path = file.path(path_heatmap, "TIL10/Monocyte/" ),
  lm22_path = file.path(path_heatmap, "LM22/Monocyte/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","Monocyte"))

cf_tbs$dataset = "Tabula Sapiens data"
cf_Liu2021$dataset = "Liu 2021 Cell data"
cf = rbind(cf_tbs,cf_Liu2021)

comp = build_confusion_heatmap(cf,  
                               c("Other_cells","Monocyte"),
                               filter = FALSE,
                               direction_color_palette = 1)
comp$plot + theme_dark() + 
  facet_wrap(. ~ dataset+subtitle, ncol=3) + 
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),

        
        plot.title = element_text(hjust = 0.5, size = 16,face="bold"))+
        
  ggtitle("Signature matrix prediction performance with Random Forest - Monocyte Classification") 
```

## Macrophage vs other

**68 Features selected**

```{r LFC_macro, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_macro = groupwise_mean[str_detect(colnames(groupwise_mean),"Macrophage")]
FC_df_macro = Get_value_RA(FC_df_macro, expr_median)
FC_df_macro = FC_prop_expr(FC_df_macro)


elbow=kneedle_threshold(FC_df_macro, FC_df_macro$FC_Macrophage, sens=1, min=1) 
elbow$plot
macro = Rprop_plot(FC_df_macro,FC_x = "FC_Macrophage", y="Macrophage_prop_expr", 
                title = "Log Fold Change of Macrophage vs other depending on the proportion of Macrophage cells expressing the genes",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

macro$plot


macro_new = row.names(macro$df[macro$df$selection != "Not selected",])


writeLines(macro_new, file.path(data_dir,'Macrophage_369.txt') )
writeLines(macro_new, file.path(git_geneset,'Macrophage_369.txt') )

```

### 369 features

```{r macro_barplot, fig.width=12, fig.height=14,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Macrophage/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_macro = Permutation_list$`RF_permutation_First_Macrophage vs other`
other_macro$genes = row.names(other_macro)

macro_computation = importance_barplot(other_macro, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 369 genes selected in Other vs Macrophage classification")
                                        
macro_computation$plot + 
  theme(axis.text.y = element_text(size=7))


macro_top68 = macro_computation$ordered_permutation_dataframe[macro_computation$ordered_permutation_dataframe$mean_importance > 0,]$genes

writeLines(macro_top68, file.path(data_dir,'Macrophage_68.txt') )
writeLines(macro_top68, file.path(git_geneset,'Macrophage_68.txt') )

```

### Performances - No Macrophages in Liu2021 data

```{r heatmap_macro, fig.width=14,fig.height=6,message=FALSE}
path_heatmap = file.path(path,"results/Tabula_sapiens/Signature_matrix_v2/")

cf_tbs = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/Macrophage/" ),
  til10_path = file.path(path_heatmap, "TIL10/Macrophage/" ),
  lm22_path = file.path(path_heatmap, "LM22/Macrophage/" ),
  pattern2= "/confusion_heatmap_",
  cells = c("Other_cells","Macrophage"))

cf_tbs$dataset = "Tabula Sapiens data"

comp = build_confusion_heatmap(cf_tbs,  
                               c("Other_cells","Macrophage"),
                               filter = FALSE,
                               direction_color_palette = 1)
comp$plot + theme_dark() + 
  facet_wrap(. ~ dataset+subtitle, ncol=3) + 
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),

        
        plot.title = element_text(hjust = 0.5, size = 16,face="bold"))+
        
  ggtitle("Signature matrix prediction performance with Random Forest - Macrophage Classification") 
```

## Platelet vs other

**40 Features selected**

```{r LFC_platelet, fig.width=13, fig.height=10,  message=F, warning=F}

row.names(groupwise_mean)= groupwise_mean$genes
FC_df_plat = groupwise_mean[str_detect(colnames(groupwise_mean),"Platelet")]
FC_df_plat = Get_value_RA(FC_df_plat, expr_median)
FC_df_plat = FC_prop_expr(FC_df_plat)

# list_elbow = list()
# for (i in seq(0,100,0.5)) {
#   el = kneedle_threshold(FC_df_plat, FC_df_plat$FC_Platelet,sens = i) 
#   name = paste0("sensitivity:",as.character(i))
#   
#   list_elbow[[name]] = el$threshold }
# df_elbow=as.data.frame(t(as.data.frame(list_elbow)))
# df_elbow$sensitivity = seq(0,100,0.5)
# 
# ggplot(data=df_elbow, aes(x=sensitivity, y=V1))+
#   geom_point() + geom_line()


elbow=kneedle_threshold(FC_df_plat, FC_df_plat$FC_Platelet,sens = 1) 
elbow$plot
plat = Rprop_plot(FC_df_plat,FC_x = "FC_Platelet", y="Platelet_prop_expr", 
                title = "Log Fold Change of Platelet vs other depending on the proportion of Platelet cells expressing the genes",
                add_label = TRUE, FC_threshold = 1, 
                prop_threshold = 0.25, 
                FC_thresh2 = elbow$threshold) 

plat$plot


platelet_new = row.names(plat$df[plat$df$selection != "Not selected",])


writeLines(platelet_new, file.path(data_dir,'Platelet_510.txt') )
writeLines(platelet_new, file.path(git_geneset,'Platelet_510.txt') )

```

### 510 features

```{r platelet_barplot, fig.width=12, fig.height=12,warning=FALSE , message=FALSE}

Permutation_list = get_permutation_list(
  directory = file.path(result_dir, "Platelet/" ),
  pattern2= "/RF_permutation/Permutation_result_")

other_plat = Permutation_list$`RF_permutation_First_Platelet vs other`
other_plat$genes = row.names(other_plat)

plat_computation = importance_barplot(other_plat, 50, 
                                        filter = "mean", 
                                        title = "Mean feature importance of 50 permutations with 510 genes selected in Other vs Macrophage classification")
                                        
plat_computation$plot + 
  theme(axis.text.y = element_text(size=10))


plat_top40 = plat_computation$ordered_permutation_dataframe[plat_computation$ordered_permutation_dataframe$mean_importance > 0,]$genes

writeLines(plat_top40, file.path(git_geneset,'Platelet_40.txt') )
writeLines(plat_top40, file.path(data_dir,'Platelet_40.txt') )

```

### Performances - Platelet unique to Tabula Sapiens

```{r heatmap_platelet, fig.width=10,fig.height=9,message=FALSE}

path_heatmap = file.path(path,"results/Tabula_sapiens/Signature_matrix_v2/")

cf_tbs = confusion_matrix_setup(
    directory = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/Platelet/" ),
    pattern2= "/confusion_heatmap_",
    cells = c("Other_cells","Platelet"),
    signature_matrix = "Tabula Sapiens")

cf_tbs$dataset = "Tabula Sapiens data"

comp = build_confusion_heatmap(cf_tbs,  
                               c("Other_cells","Platelet"),
                               filter = FALSE,
                               direction_color_palette = 1)
comp$plot + theme_dark() + 
  facet_wrap(. ~ dataset+subtitle, ncol=3) + 
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),
        plot.title = element_text(hjust = 0.5, size = 16,face="bold"))+
        
  ggtitle("Signature matrix prediction performance with Random Forest - Platelet Classification") 
```

# Combining Features

```{r combine_features3.2, fig.width=16, fig.height=10,warning=FALSE , message=FALSE}


LM22 = read.table(file.path(path,"results/Tabula_sapiens/LM22/LM22_merged.txt"), sep="\t")
TIL10 = read.table(file.path(path,"results/Tabula_sapiens/LM22/TIL10_binary_standardized.txt"), sep="\t")


LM22_list = lapply( names(LM22), function(x) {
  geneset = row.names(LM22[LM22[[x]] ==1 ,])
  return(geneset) })
names(LM22_list) = colnames(LM22)

TIL10_list = lapply( names(TIL10), function(x) {
  geneset = row.names(TIL10[TIL10[[x]] ==1 ,])
  return(geneset) })
names(TIL10_list) = colnames(TIL10)


TabulaSapiens_geneset = list(
  "NK_cell"   = unique(c(top79_nkcd8_genes,top178_nk)),
  "Neutrophil"= neutro_top84,
  "Monocyte"  = mono_top174,
  "Plasmocyte"= plasmo_top19,
  
  "CD4"       = unique(c(CD4_top114, top61_cd48_genes)),
  "CD8"       = unique(c(top61_cd48_genes, top79_cd8, top247_nkcd8_genes)), 
  
  "Macrophage"= macro_top68,
  "Platelet"  = plat_top40,
  "memory_B_cell"   = unique(c(top42_naive_mem, top147_memB)), 
  "naive_B_cell"    = unique(c(naiveB_top17,top42_naive_mem))
  )

save(TabulaSapiens_geneset, file = file.path(git_geneset,"TabulaSapiens_geneset_final_erratum.Rdata"))
  
geneset_flat = unique(unlist(TabulaSapiens_geneset))


upset(fromList(TabulaSapiens_geneset),     
      sets.bar.color = "#56B4E9", order.by = "freq", nsets = 50,
      empty.intersections = NULL, set_size.show = TRUE, mb.ratio = c(0.7, 0.3),
      text.scale = 1.75, set_size.scale_max = 400)

plot_list = list()
for (cell in names(TabulaSapiens_geneset)) {
  if (cell != "Platelet" ){
  
  cell_list = list(
    TabulaSapiens_geneset[[cell]],
    TIL10_list[[cell]], 
    LM22_list[[cell]] )
  
  names(cell_list) = c(paste0("Tabula Sapiens - ",cell),
                       paste0("TIL10 - ",cell), 
                       paste0("LM22 - ",cell) )
  
  plot_list[[cell]]= upset(fromList(cell_list),     
      sets.bar.color = "#56B4E9", order.by = "freq", nsets = 50,
      empty.intersections = NULL, set_size.show = TRUE, mb.ratio = c(0.7, 0.3),
      text.scale = 1.75, set_size.scale_max = max(unlist(lapply(cell_list, function(x) length(x)))) * 1.25 )
}}

plot_list


```

### Performances

```{r heatmap_liu, fig.width=16,fig.height=6,message=FALSE}
 path_heatmap = file.path(path,"results/Liu2021/Signature_matrix_v2/")
cells = c("CD4","CD8","Monocyte","NK_cell","memory_B_cell","naive_B_cell")

cf_Liu2021 = merge_confusion_matrix( 
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/Multiclass/" ),
  til10_path = file.path(path_heatmap, "TIL10/Multiclass/RF/" ),
  lm22_path = file.path(path_heatmap, "LM22/Multiclass/RF/" ),
  pattern2= "/confusion_heatmap_",
  cells = cells)

path_heatmap = file.path(path,"results/Tabula_sapiens/Signature_matrix_v2/")
cf_Liu2021$dataset = "Liu 2021 Cell data"

comp = build_confusion_heatmap(cf_Liu2021,
                               c("NK_cell","CD8",
                                 "CD4","Monocyte",
                                 "memory_B_cell","naive_B_cell"),
                               filter = FALSE,
                               direction_color_palette = -1)

comp$plot + theme_dark() + 
  facet_wrap(. ~ dataset+subtitle, ncol=3) + 
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        plot.title = element_text(hjust = 0.5, size = 16,face="bold"))+
        
  ggtitle("Signature matrix prediction performance with Random Forest - Multiclass Classification") 
```

```{r heatmap_tbs, fig.width=18,fig.height=8,message=FALSE}
cells = c("CD4","CD8","Erythrocyte","Macrophage",
          "Monocyte","NK_cell","Neutrophil", 
          "memory_B_cell","naive_B_cell")
path_heatmap = file.path(path,"results/Tabula_sapiens/Signature_matrix_v2/")

cf_tbs = merge_confusion_matrix(
  tbs_path = file.path(path_heatmap, "Tabula_Sapiens_50k_erratum/Multiclass/" ),
  til10_path = file.path(path_heatmap, "TIL10/Multiclass/" ),
  lm22_path = file.path(path_heatmap, "LM22/Multiclass/" ),
  pattern2= "/confusion_heatmap_",
  cells = cells)

cf_tbs$dataset = "Tabula Sapiens data"

cf_tbs = cf_tbs[ cf_tbs$True_label != "Erythrocyte" &  cf_tbs$Predicted_label != "Erythrocyte",]

balanced_acc1 = round(mean(cf_tbs[cf_tbs$True_label == cf_tbs$Predicted_label & cf_tbs$signature_matrix == "Tabula Sapiens",]$Percentage)*100,2)

balanced_accLM22 = round(mean(cf_tbs[cf_tbs$True_label == cf_tbs$Predicted_label & cf_tbs$signature_matrix == "LM22",]$Percentage)*100,2)

balanced_accTIL10 = round(mean(cf_tbs[cf_tbs$True_label == cf_tbs$Predicted_label & cf_tbs$signature_matrix == "TIL10",]$Percentage)*100,2)


cf_tbs[cf_tbs$signature_matrix == "Tabula Sapiens",]$subtitle = 
  paste0("Tabula Sapiens \n Balanced accuracy = ",balanced_acc1,"%")

cf_tbs[cf_tbs$signature_matrix == "TIL10",]$subtitle = 
  paste0("TIL10 \n Balanced accuracy = ",balanced_accTIL10,"%")

cf_tbs[cf_tbs$signature_matrix == "LM22",]$subtitle = 
  paste0("LM22 \n Balanced accuracy = ",balanced_accLM22,"%")
  
  
comp = build_confusion_heatmap(cf_tbs,
                               c("NK_cell","CD8","CD4",
                                 "Monocyte","Macrophage","Neutrophil",
                                 "memory_B_cell","naive_B_cell"),
                               filter = FALSE,
                               direction_color_palette = -1)
comp$plot + theme_dark() + 
  facet_wrap(. ~ dataset+subtitle, ncol=3) + 
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        plot.title = element_text(hjust = 0.5, size = 16,face="bold") )+
        
  ggtitle("Signature matrix prediction performance with Random Forest - Multiclass Classification") 
```

```{r combine_features3.4, fig.width=16, fig.height=10, include=FALSE, eval=FALSE}
TabulaSapiens_binary = geneset_to_binary_df(TabulaSapiens_geneset)
print(table(rowSums(TabulaSapiens_binary)))

writeLines(geneset_flat, file.path(path,"/results/Tabula_sapiens/LM22/signature_genes_TabulaSapiens_50k_erratum.txt"))

write.table(TabulaSapiens_binary, file.path(path,"/results/Tabula_sapiens/LM22/TabulaSapiens_binary_erratum.txt"),
             quote = FALSE, row.names = TRUE, sep='\t')
            

```

# Mean Expression heatmap

```{r mean_expr_heat630, fig.width=20, fig.height=12, include = FALSE,eval=TRUE}

heatmap = groupwise_mean[groupwise_mean$genes %in% geneset_flat,]
heatmap = heatmap[,! str_detect(colnames(heatmap), "other|Erythrocyte|prop_expr")]


row.names(heatmap) = heatmap$genes
heatmap$genes = NULL


mat=read.csv(paste0(path,"/data/Tabula_sapiens_immune_all/raw_Tabula_sapiens_1039genes.txt"), 
             sep='\t',row.names = 1, header = TRUE)


#####
cellTypes = readLines(file.path(data_dir,"Celltypes_of_interest_50k.txt"))
mat$cell_id = cellTypes

# Aggregate rows by group by using mean function
groupwise_mean = aggregate(mat[colnames(mat)[colnames(mat) != 'cell_id']], list(mat$cell_id), FUN=mean)
groupwise_mean = t(groupwise_mean)
colnames(groupwise_mean) = groupwise_mean[1,]
groupwise_mean = groupwise_mean[-1,]
summary(groupwise_mean)

X <- as.data.frame(apply(groupwise_mean, 2, as.numeric))
row.names(X) = row.names(groupwise_mean)
X$other = NULL

write.table(X, file.path(data_dir,"Signature_TabulaSapiens_1039_raw.txt"),
           quote = FALSE, row.names = TRUE, sep='\t')
#####


write.table(heatmap, file.path(data_dir,"Signature_TabulaSapiens_1039.txt"),
           quote = FALSE, row.names = TRUE, sep='\t')
 write.table(heatmap, file.path(git_dir,"/Signature_matrix/Signature_TabulaSapiens_1039_10celltypes.txt"),
           quote = FALSE, row.names = TRUE, sep='\t')

```



```{r mean_expr_heat, fig.width=20, fig.height=10,warning=FALSE , message=FALSE}


genes_annotation = heatmap
genes_annotation$NK_cell = ifelse(row.names(genes_annotation) %in% TabulaSapiens_geneset$NK_cell, "yes" , "no")
genes_annotation$CD4 = ifelse(row.names(genes_annotation) %in% TabulaSapiens_geneset$CD4, "yes" , "no")
genes_annotation$CD8 = ifelse(row.names(genes_annotation) %in% TabulaSapiens_geneset$CD8,"yes" , "no")
genes_annotation$Neutrophil = ifelse(row.names(genes_annotation) %in% TabulaSapiens_geneset$Neutrophil,"yes" , "no")
genes_annotation$naive_B_cell = ifelse(row.names(genes_annotation) %in% TabulaSapiens_geneset$naive_B_cell,"yes" , "no")
genes_annotation$memory_B_cell = ifelse(row.names(genes_annotation) %in% TabulaSapiens_geneset$memory_B_cell,"yes" , "no")
genes_annotation$Plasmocyte = ifelse(row.names(genes_annotation) %in% TabulaSapiens_geneset$Plasmocyte,"yes" , "no")
genes_annotation$Monocyte = ifelse(row.names(genes_annotation) %in% TabulaSapiens_geneset$Monocyte,"yes" , "no")
genes_annotation$Macrophage = ifelse(row.names(genes_annotation) %in% TabulaSapiens_geneset$Macrophage,"yes" , "no")
genes_annotation$Platelet = ifelse(row.names(genes_annotation) %in% TabulaSapiens_geneset$Platelet,"yes" , "no")


color_yes = "darkgreen"
color_no = "#CCCCCC"

my_colour = list(
  B_cell = c(yes = color_yes, no = color_no),
  naive_B_cell = c(yes = color_yes, no = color_no),
  memory_B_cell = c(yes = color_yes, no = color_no),
  Neutrophil = c(yes = color_yes, no = color_no),
  CD4 = c(yes = color_yes, no = color_no), 
  NK_cell = c(yes = color_yes, no = color_no) ,
  CD8 = c(yes = color_yes, no = color_no),
  Plasmocyte = c(yes = color_yes, no = color_no),
  Monocyte = c(yes = color_yes, no = color_no),
  Macrophage = c(yes = color_yes, no = color_no),
  Platelet = c(yes = color_yes, no = color_no) )

setHook("grid.newpage", function() pushViewport(viewport(x=1,y=1,width=1, height=0.95, name="vp", just=c("right","top"))), action="prepend")
pheatmap(t(heatmap), scale = "column", 
         cutree_cols = 10,
         angle_col = "45", main = "Mean expression heatmap of the 1039 features used to predict 10 cell types",
         clustering_distance_rows = 'correlation', 
         clustering_distance_cols = 'correlation',
         clustering_method = "ward.D2",
         annotation_col = genes_annotation, annotation_colors = my_colour,
         color = colorRampPalette(c("navy", "#DDDDDD", "firebrick"))(25),
         show_colnames = FALSE)
setHook("grid.newpage", NULL, "replace")
grid.text("Genes", y=-0.02, gp=gpar(fontsize=16))


```

```{r mean_expr_heat2, fig.width=11, fig.height=15,warning=FALSE , message=FALSE, eval=FALSE, include=FALSE}

setHook("grid.newpage", function() pushViewport(viewport(x=1,y=1,width=1, height=1, name="vp", just=c("right","top"))), action="prepend")
pheatmap(heatmap, scale = "row", 
         cutree_rows = 10,
         angle_col = "45", main = "Mean expression heatmap of the 1039 features used to predict 10 cell types",
         clustering_distance_rows = 'correlation', 
         clustering_distance_cols = 'correlation',
         clustering_method = "ward.D2",
         annotation_row = genes_annotation, annotation_colors = my_colour,
         color = colorRampPalette(c("navy", "#DDDDDD", "firebrick"))(25),
         show_rownames = FALSE)
setHook("grid.newpage", NULL, "replace")
grid.text("Genes", y=0.35, x=0.9, gp=gpar(fontsize=16))
```


# ROC curve


```{r ROC_curve, fig.width=13,fig.height=8}
perf_dir_Liu = file.path(path, "results/Liu2021/Signature_matrix_v2/") 
files = list.files(path = perf_dir_Liu, pattern = "ROC_values",  recursive = TRUE)
files = files[!str_detect(files, "/LR")]


ROC = lapply(files, function(f) {
  table = read.csv2(file.path(perf_dir_Liu,f), header = TRUE, sep = ",", row.names = 1) } )

names(ROC) = files

signature = unname(sapply(files, function(x) str_split(x, "/")[[1]][1]))
cell = unname(sapply(files, function(x) str_split(x, "/")[[1]][2]))


for (index in seq(1:length(ROC))){
  df = ROC[[index]]
  df = cbind(df, cell[index])
  df = cbind(df, signature[index])
  colnames(df)[(length(colnames(df))-1) : length(colnames(df))  ] = c("cell","signature")
  ROC[[index]] = df
}


ROC_df = ldply(ROC)
ROC_df$clf = NULL
ROC_df$.id = NULL

ROC_df = ROC_df[ROC_df$signature != "Tabula_Sapiens_50k" , ]
ROC_df$signature = ifelse(ROC_df$signature == "Tabula_Sapiens_50k_erratum", 
                          yes = "Tabula_Sapiens_50k", no = ROC_df$signature )

ROC_df$auc_label = paste0(ROC_df$signature," - AUC = ",as.character(round(as.numeric(ROC_df$auc),4) ))



annotation_df = ROC_df
annotation_df$fpr = NULL
annotation_df$tpr = NULL
annotation_df = unique(annotation_df)

annotation_df$y = ifelse(annotation_df$signature == "LM22", 0.3, 0.05 )
annotation_df$y = ifelse(annotation_df$signature == "Tabula_Sapiens_50k", 0.175, annotation_df$y )


ggplot(ROC_df, aes(x=as.numeric(fpr),y=as.numeric(tpr),color=signature) ) + 
  geom_line(alpha=1) +
  geom_abline(linetype = 2, size = 0.25) + 
  theme_linedraw() + 
  labs(color = " ") +
  ggtitle("Celltypes ROC curves for the Liu et al, 2021, Cell dataset - Signature matrix comparison")+
  xlab("False Positive Rate") + ylab("True Positive Rate") +
  geom_label(data=annotation_df, aes(x=1,y=y,label=auc_label,color=signature), size = 3, hjust = 1) + 
  facet_wrap(. ~ cell, ncol=3) +
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),
        plot.title = element_text(hjust = 0.5))

```


```{r ROC_curve2, fig.width=13,fig.height=9}
perf_dir_Liu = file.path(path, "results/Tabula_sapiens/Signature_matrix_v2/") 
files = list.files(path = perf_dir_Liu, pattern = "ROC_values",  recursive = TRUE)
files = files[!str_detect(files, "/LR")]


ROC = lapply(files, function(f) {
  table = read.csv2(file.path(perf_dir_Liu,f), header = TRUE, sep = ",", row.names = 1) } )

names(ROC) = files

signature = unname(sapply(files, function(x) str_split(x, "/")[[1]][1]))
cell = unname(sapply(files, function(x) str_split(x, "/")[[1]][2]))


for (index in seq(1:length(ROC))){
  df = ROC[[index]]
  df = cbind(df, cell[index])
  df = cbind(df, signature[index])
  colnames(df)[(length(colnames(df))-1) : length(colnames(df))  ] = c("cell","signature")
  ROC[[index]] = df
}


ROC_df_tbs = ldply(ROC)
ROC_df_tbs$clf = NULL
ROC_df_tbs$.id = NULL

ROC_df_tbs$auc_label = paste0(ROC_df_tbs$signature," - AUC = ",as.character(round(as.numeric(ROC_df_tbs$auc),4) ))

ROC_df_tbs = ROC_df_tbs[ROC_df_tbs$signature != "Tabula_Sapiens_50k-10celltypes",]

ROC_df_tbs = ROC_df_tbs[ROC_df_tbs$signature != "Tabula_Sapiens_50k" , ]
ROC_df_tbs$signature = ifelse(ROC_df_tbs$signature == "Tabula_Sapiens_50k_erratum", 
                          yes = "Tabula_Sapiens_50k", no = ROC_df_tbs$signature )

annotation_df = ROC_df_tbs
annotation_df$fpr = NULL
annotation_df$tpr = NULL
annotation_df = unique(annotation_df)

annotation_df$y = ifelse(annotation_df$signature == "LM22", 0.3, 0.05 )
annotation_df$y = ifelse(annotation_df$signature == "Tabula_Sapiens_50k", 0.175, annotation_df$y )



ggplot(ROC_df_tbs, aes(x=as.numeric(fpr),y=as.numeric(tpr),color=signature) ) + 
  geom_line(alpha=1) +
  geom_abline(linetype = 2, size = 0.25) + 
  theme_linedraw() + 
  labs(color = " ") +
  xlab("False Positive Rate") + ylab("True Positive Rate") +
  geom_label(data=annotation_df, aes(x=1,y=y,label=auc_label,color=signature), size = 3, hjust = 1) + 
  facet_wrap(. ~ cell, ncol=4) +
  ggtitle("Celltypes ROC curves for the Tabula Sapiens dataset - Signature matrix comparison")+
  theme(strip.text = element_text(size=14,face="bold"),
        panel.background  = element_rect(fill = "white"),
        panel.grid.major  = element_line(color = "white"),
        panel.border = element_rect(color = "black",fill=NA),
        legend.position = c(0.75,0.15),
        plot.title = element_text(hjust = 0.5))

```




```{r save_R, eval=FALSE,include=FALSE}
rmd2rscript <- function(infile){
  # read the file
  flIn <- readLines(infile)
  # identify the start of code blocks
  cdStrt <- which(grepl(flIn, pattern = "```{r*", perl = TRUE))
  # identify the end of code blocks
  cdEnd <- sapply(cdStrt, function(x){
    preidx <- which(grepl(flIn[-(1:x)], pattern = "```", perl = TRUE))[1]
    return(preidx + x)
  })
  # define an expansion function
  # strip code block indacators
  flIn[c(cdStrt, cdEnd)] <- ""
  expFun <- function(strt, End){
    strt <- strt+1
    End <- End-1
    return(strt:End)
  }
  idx <- unlist(mapply(FUN = expFun, strt = cdStrt, End = cdEnd, 
                SIMPLIFY = FALSE))
  # add comments to all lines except code blocks
  comIdx <- 1:length(flIn)
  comIdx <- comIdx[-idx]
  for(i in comIdx){
    flIn[i] <- paste("#' ", flIn[i], sep = "")
  }
  # create an output file
  nm <- strsplit(infile, split = "\\.")[[1]][1]
  flOut <- file(paste(nm, "[rmd2r].R", sep = ""), "w")
  for(i in 1:length(flIn)){
    cat(flIn[i], "\n", file = flOut, sep = "\t")
  }
  close(flOut)
}

rmd2rscript("/home/acolajanni/Documents/work/results/Tabula_sapiens/markdown/Feature_selection_50k.Rmd")
```
